Enabling Stretch Kafka Deployments with Strimzi (#129)
strimzi/proposals: 095-stretch-cluster.md

1. Valid broker and controller service endpoints values for `advertised.listeners` and `controller.quorum.voters` within the appropriate ConfigMap resources.
2. Broker certificates with appropriate Subject Alternative Names.
The prototype currently expects this identifier to be supplied by the user as a label value on the Kafka node pool resource:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  ...
  labels:
    ...
    strimzi.io/submariner-cluster-id: "cluster1"
```

_Note: The prototype operator code currently expects `submariner-cluster-id`, but this is subject to change for something more generic once additional technologies have been prototyped. One suggestion is `strimzi.io/cluster-network-id`._

_Note: This network identifier could have the same value as the Kubernetes cluster identifier described in the following section and both aspects would benefit from community feedback and input._

#### User configuration of the deployment topology

The operator running in the central Kubernetes cluster must be provided with the following information to allow creation of resources in remote clusters:

- An identifier for each remote Kubernetes cluster.

> - A URL endpoint for the Kubernetes API server running in each remote cluster.

Reviewer: Would you expect for this traffic to go across the cloud native network?

Author: Yes, for pod to pod communication between Kafka brokers and controllers across k8s clusters we are using cloud native networking solutions like Submariner, Cilium . However when the central Operator needs to create resource in remote/Member clusters (for eg: SPS) , it communicates directly with k8s API of the target cluster using credentials stored in secrets. this approach ensures that resource management remains centralized while keeping the cross cluster pod communication efficient . as of now, this is the plan

- Credential(s) to allow authentication with the remote Kubernetes API servers.

For simplicity and to quickly achieve a working prototype, the information outlined above is currently provided as an environment variable to the central operator.
The value of the environment variable uses a "map" format as shown below:

```yaml

> - name: STRIMZI_K8S_CLUSTERS

Reviewer: Is it better to have cluster credentials at the operator level, or would it make more sense to have the credentials for each KafkaNodePool?

Perhaps the KNP `spec.cluster` could be:
```
spec:
  cluster:
    secretName: <my-cluster-secret-with-kubecfg-in-it>
```

Reviewer 2: Sooner or later, you will likely need to have some watches / informers for the clusters etc. So you need to know up-front what clusters you will manage to set those up. So I do not think this should be passed as the operand configuration, it needs to be operator configuration

Author: Thanks for your feedback @Jamstah @scholzj , I agree that the operator needs to know all managed clusters upfront to setup the necessary watches and informers. defining clusters at the operator level ensures that we can initilize informers at startup rather than trying to dynamically adjust them as KNPs are added. the env var approach was chosen to quickly prototype this, but we can explore a CRD based approach in the future if needed. That said, keeping  cluster creds centralized also help with consistency and security management , rather distributing them across multiple KNPs

The Strimzi cluster operator will use this identifier in two ways:
1. Generating `advertised.listeners` and `controller.quorum.voters` configuration.
2. Determining the target Kubernetes cluster for a Kafka node pool.

#### Step 2: Deploy a cluster operator to each Kubernetes cluster

##### Central cluster operator configuration

The operator running in the central Kubernetes cluster must be provided with the following information to allow creation of resources in remote clusters:

- The identifier for each Kubernetes cluster defined in 'Step 1'.
- A URL endpoint for the Kubernetes API server running in each remote cluster.
- Credential(s) to allow authentication with the remote Kubernetes API servers.

The information outlined above will be provided as an environment variable for the cluster operator.
The value of the environment variable uses a "map" format as shown below:

```yaml
- name: STRIMZI_K8S_CLUSTERS
  value: |
      cluster-id-a.url=<cluster-a URL>
      cluster-id-a.secret=<secret-name-cluster-a>
      cluster-id-b.url=<cluster-b URL>
      cluster-id-b.secret=<secret-name-cluster-b>
```

> The secrets referenced here must contain the kubeconfig for the Kubenetes cluster available at the provided URL as the value of secret key 'kubeconfig'.

Reviewer 3: Can we put an example here of such a Secret?

Author: Added the example

In addition, Strimzi also creates network policies to restrict traffic from outside the namespace.
To enable pod-to-pod communication across clusters, the cluster operator will not create any network policies.

Furthermore, depending on the listener configuration, the Kafka CR may also create LoadBalancer, Ingress, NodePort, and Route services to expose Kafka externally.
By ensuring that all necessary resources including ConfigMaps, Secrets, Services, and PVCs—are created in the appropriate clusters, the central operator enables  Kafka deployment across multiple Kubernetes clusters in a Stretch Kafka setup.

Operators deployed to remote clusters are only responsible for reconciling `StrimziPodSet` resources that are created remotely by the operator running in the central cluster.

This approach will allow users to manage the definition of their stretch Kafka cluster in a single location.

#### Prototype

A working prototype can be deployed using the steps outlined in a draft [README](https://aswinayyolath.github.io/stretch-kafka-docs/) that is being iteratively revised.

> _Note: The prototype might not exactly align with this proposal so please refer to the `README` documentation when working with the prototype._

Reviewer 3: This is dangerous because it sounds like what you are proposing could be different from what you are implementing, so which one would be the source of truth?

Author: We will be updating the prototype code also to match with the design. The revised proposal IMO will address most of the early confusions. Now we will take a look at the un addressed review comments and make necessary revisions and corrections and make sure this will reflect in the prototype as well

Reviewer 4: We have discussed this further and have decided not to update the prototype to match the proposed design. Instead, we plan to put together a number of PRs (as discussed during the last community call) that will be the "source of truth" for the implementation.

Reviewer 3: I totally agree on having multiple PRs to simplify the reviews (of course only when the proposal will be accepted). Not clear why not updating the POC accordingly if someone would like to try it. Of course, having the POC is not a requirement for the proposal so it's up to you.

Reviewer 4: We are working on an implementation based on the updated design reflected in the proposal. We are planning to make that code available to try once it is ready. This will likely be a new branch on a public fork.

Our plan is to raise draft PRs that together will provide a stretch cluster implementation to allow for the community to review as work progresses.

Our previous prototype was based on Strimzi 0.44 and we are now implementing on the latest code from the Strimzi `main` branch.

### Additional considerations and reference information

#### Authentication & Security Considerations for Multi-Cluster Access

The Strimzi operator itself remains agnostic to the authentication method used; It simply consumes the kubeconfig provided by the user. 
The responsibility for securing credentials, whether using tokens, certificates, or other mechanisms, lies with the user.

> We have provided additional details on securing remote cluster credentials in a multi-cluster Kafka deployment [here](https://aswinayyolath.github.io/stretch-kafka-docs/Secure-remote-cluster-credentials/). 

Reviewer 3: Why are these details not part of the proposal? My fear is that we are spreading the content across two sources, or?

Author: We kept the proposal focused on the core architecture and operator behavior, while the linked document provides an example of securing credentials specifically for token-based authentication. The reason for this separation is that authentication methods can vary (e.g., certificates, tokens, OIDC), and we didn’t want to prescribe a single approach within the proposal.

That said, if you feel it's important to have some of these details within the proposal itself for completeness, we can include a brief summary while keeping implementation specifics in the linked documentation. Let us know what you think!

Reviewer 3: I think that an overall view for supported authentication methods would be beneficial to the proposal. I would like people to jump to the POC only if they want to try it but not for understanding. The overall concepts should be in the proposal to be self-consistent.

Author: Okay that make sense... I will update this section accordingly

```
controller.quorum.voters=<controller-id>@<controller-pod-name>.<stretch-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local:9090, ...
```

These updates ensure brokers and controllers can be discovered and communicate across clusters without relying on traditional external access methods.

#### Resource cleanup on remote Kubernetes clusters

In our POC, resources created in the remote clusters do not have `OwnerReferences`. The main reason is that the `Kafka` and `KafkaNodePool` CRs exist only in the central cluster. If resources in remote clusters were to reference them as owners, they would be immediately deleted by Kubernetes' garbage collection mechanism since their owners do not exist in the remote clusters.

Even if ownership across Kubernetes cluster boundaries were possible, it would introduce another issue: if the central cluster were to go down, any resources in remote clusters with OwnerReferences pointing to central cluster resources would also be deleted automatically. This is not the intended behaviour, as we want remote resources to remain operational even if the central cluster becomes temporarily unavailable.

> The prototype does not yet have a mechanism for controlled cleanup of remote resources when a user deliberately deletes the Kafka and KafkaNodePool CRs from the central cluster. Currently, when a user deletes these CRs, all related resources in the central cluster are removed, but remote cluster resources remain. Finding a way to handle this cleanup while avoiding unintended deletions is an open challenge.

Reviewer 3: Why doesn't the central cluster delete the corresponding resources on the remote clusters as it creates them in the beginning?

Reviewer 4: We are currently working on this area and hope to have some updates next week to help answer your question and clarify the proposal here.

Reviewer 4: We have updated this section with a more precise proposal about how remote resources will be managed.

##### Recovery Approaches
There are several possible recovery methods

###### 1. Manual Recovery (Baseline Approach)
Users bring up a new Kubernetes cluster and manually apply the Kafka, KafkaNodePool CRs.
Alternatively, an existing remote cluster can be promoted to central by applying these CRs and updating the Cluster Operator deployment.
This approach is simple and reliable but requires user intervention.

###### 2. GitOps-Based Recovery

Users store CRs in GitOps tools (e.g., ArgoCD, Flux).
When a new central cluster is created, GitOps tools automatically reconcile the resources.
This reduces manual effort but requires GitOps workflows to be set up in advance.

#### Entity operator
We would recommend that all KafkaTopic and KafkaUser resources are managed from the cluster that holds Kafka and KafkaNodePool resources, and that should be the cluster where the entity operator should be enabled.
This will maintain all resource management from a central point.
The expectation is that the entity operator will not be impacted by changes made in the prototype, but this currently requires verification.

> Other aspects requiring consideration and research:

Reviewer 3: Does it mean that the following will not be addressed in this proposal, or just you need more time to think and propose how to handle them?

Reviewer 4: This section was added to highlight some areas that might need attention at a later stage. Having revisited this, we believe this section should be removed and details for any specific changes we are proposing in any of these areas included in the main proposal.

The information outlined above will be provided as an environment variable for the cluster operator.
The value of the environment variable uses a "map" format as shown below:

```yaml
- name: STRIMZI_REMOTE_KUBE_CONFIG
  value: |
      cluster-id-a.url=<cluster-a URL>
      cluster-id-a.secret=<secret-name-cluster-a>
      cluster-id-b.url=<cluster-b URL>
      cluster-id-b.secret=<secret-name-cluster-b>
```

The secrets referenced here must contain the kubeconfig for the Kubernetes cluster available at the provided URL as the value of secret key 'kubeconfig'.
This allows the central Strimzi operator to authenticate with multiple Kubernetes clusters.

**Example Secret**

Below is an example Kubernetes Secret containing a kubeconfig for a remote cluster

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: secret-name-cluster-a
  namespace: strimzi
type: Opaque
data:
  kubeconfig: <base64-encoded-kubeconfig>
```

When this environment variable is set, the cluster operator will understand it needs to deploy a stretch Kafka cluster.
The custom actions taken as a result of this environment variable include:

> 1. The value of `STRIMZI_NETWORK_POLICY_GENERATION` is assumed to be `false` irrespective of the value set by the user.

Author: You're absolutely right , To clarify, in stretch mode, the operator ignores any value set by the user for `STRIMZI_NETWORK_POLICY_GENERATION` and explicitly treats it as **false** to ensure that cross-cluster network traffic is not blocked. We've updated the proposal to reflect this behavior and mentioned that a warning will be logged to notify users that their configured value is being overridden in this mode.

This is to ensure network traffic between Kubernetes clusters is not blocked.

> 2. The `StrimziPodSet` resources created in a remote cluster will include an annotation `strimzi.io/orphaned-podset: true`.

Author: You're right, the annotation is specifically intended for the stretch cluster use case where StrimziPodSets exist in remote clusters without the accompanying Kafka and KafkaNodePool CRs. At this point, we don't envision this being a more general-purpose feature. So renaming the annotation to something like `strimzi.io/remote-podset` makes sense and better reflects its intended scope.

This annotation will allow the remote cluster operator to reconcile the `StrimziPodSet` without a `Kafka` and `KafkaNodePool` CR being present in the same cluster.

##### Remote cluster operator configuration

> The operator running in the remote cluster must be set to only reconcile `StrimziPodSet` resources by setting the existing environment variable:

Author: Thanks for raising this point! To clarify `STRIMZI_POD_SET_RECONCILIATION_ONLY` is optional. It is intended as an optimization for the remote cluster operator to reduce resource usage and simplify logging. Stretch cluster functionality will still work correctly even if this environment variable is not set.

We've updated the proposal to reflect that this is an optional configuration and not a requirement.

> The operator running in the remote cluster must be set to only reconcile `StrimziPodSet` resources by setting the existing environment variable:

Reviewer 3: You are right, good point. Even if when it will come to documenting how to deploy a strech cluster I would be more opinionated and making this step not that much optional.

> The operator running in the remote cluster must be set to only reconcile `StrimziPodSet` resources by setting the existing environment variable:

Author: we should clearly document that `STRIMZI_POD_SET_RECONCILIATION_ONLY` should be set to true when deploying the operator to remote clusters. I will update that in the proposal

```yaml
- name: STRIMZI_POD_SET_RECONCILIATION_ONLY
  value: true
```

This configuration will minimise the resources required by the cluster operator on the remote Kubernetes clusters and simplify logging output.

#### Step 3: Create Kafka and KafkaNodePool resources in the central cluster

##### Kafka CR

> The `Kafka` CR must include a new annotation to specify the chosen Cloud native network project:

Reviewer 3: can you add an explanation about why this annotation is needed. My guess is that the operator has to deal with some network project specific configuration/interaction but it would be better to give more info here.
Also, what happens if the users forgot to set it? Still one more thing to clarify.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  annotations:
    strimzi.io/stretch-cluster-network: "cilium"
```

##### KafkaNodePool CR

> The `KafkaNodePool` CR must include a new annotation to specify the target Kubernetes cluster where the node pool resources should be deployed:

Reviewer 3: I guess this would be misleading. I mean we are not creating/deploying the KafkaNodePool on the remote cluster but only the StrimziPodSet. The KNP will stay on the central cluster. This sentence made me think that KNP is going to be deployed in the remote cluster defined by the annotation instead.

Author: Thanks for the feedback, you're absolutely right that the original wording was misleading. To clarify: the KafkaNodePool remains in the central cluster, and only the dependent Kubernetes resources (such as StrimziPodSet, Services, Secrets, ConfigMaps, etc.) are created in the remote cluster specified by the annotation. We've updated the proposal to make this distinction clearer.

##### Kafka CR

To enable stretch cluster functionality, the `Kafka` CR must include a new annotation to specify the cloud-native networking technology used for cross-cluster communication:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  annotations:
    strimzi.io/stretch-cluster-network: "cilium"
```

This annotation is not used to set up cross-cluster networking — that remains the responsibility of the user or platform. 
Instead, it allows the Cluster Operator to apply any technology-specific configuration or integration needed for the stretch cluster to function correctly.
For example, when using Submariner, the operator may need to automatically create `ServiceExport` resources to enable service discovery across clusters.
Other technologies like Cilium do not require such configuration.

The annotation also makes the implementation future-proof: as support for more networking technologies is added, this field allows the operator to adapt its behavior accordingly.

If the annotation is omitted, the operator will assume "cilium" as the default, as it does not require additional configuration.

> A validating webhook will be used to ensure the annotation, if present, contains a supported value.

Reviewer 3: Tbh I am not sure we want to add a webhook to our Strimzi cluster operator for several reasons:

* it adds an external dependency. If the webhook doesn't work for any reasons, the entire admission process fails, and it means that your resource won't be created and your cluster won't be up and running.
* webhooks need installation, also security via TLS certificates ... it means more operational burden
* can't think of debugging and testing issues as well
* an additional point of compatibility risk when it comes to upgrade from an operator version to another

So in general I would be more for the operator itself to validate the presence or value of such an annotation and dealing with it properly, i.e. logging and failing the reconciliation.
But of course, other @strimzi/maintainers could have a different opinion.

Author: Thanks for the feedback! After giving it more thought, I believe it may actually be more beneficial to move the stretch-cluster-network configuration from the Kafka CR annotation to an environment variable in the Cluster Operator deployment. 

for example

```yaml
- name: STRIMZI_STRETCH_CLUSTER_NETWORK
  value: "cilium" # or any other tech that is supported in future 
```

Here are the main reasons why I think this makes sense

- We already configure cross-cluster support (like remote cluster credentials) through the `STRIMZI_REMOTE_KUBE_CONFIG` environment variable. Defining the networking technology (e.g., Cilium, Submariner) in the same place creates a single source of truth and keeps all stretch-related configuration tightly scoped to the operator.
- The Kafka resource can remain focused on describing the Kafka deployment itself, without embedding operator/platform-specific configuration. This separation of concerns can make the CRs easier to read and maintain.
- This approach reduces the chance of misconfiguration —>> users don't need to remember to add specific annotations to Kafka CR. It’s easier to document, automate, and debug when all stretch-specific config lives in one place.
- Since we only support one networking technology per Stretch cluster deployment, having this configured at the operator level aligns well with the current design.

This would also avoid the need for admission webhooks and sidestep the operational complexity that comes with them, while still allowing the operator to validate and respond appropriately to invalid or unsupported values.

Happy to hear what you think — and open to further discussion, of course

Author: Hi @ppatierno, @katheris, @fvaleri,

I’ve updated this section based on my earlier comment - replacing the annotation in the Kafka CR with an env var in the CO deployment. I believe this revised design is a cleaner and more robust solution than what we originally proposed. Could you please take a look?

Reviewer 3: I was going to suggest the same, so I agree. It's more of an operator specific configuration than a Kafka cluster one.

Author: Thanks, I have updated the proposal to address this, I will mark this thread resolved after your review

- A URL endpoint for the Kubernetes API server running in each remote cluster.
- Credential(s) to allow authentication with the remote Kubernetes API servers.

For simplicity and to quickly achieve a working prototype, the information outlined above is currently provided as an environment variable to the central operator.
The value of the environment variable uses a "map" format as shown below:

```yaml
- name: STRIMZI_K8S_CLUSTERS
  value: |
      cluster-id-a.url=<cluster-a URL>
      cluster-id-a.secret=<secret-name-cluster-a>
      cluster-id-b.url=<cluster-b URL>
      cluster-id-b.secret=<secret-name-cluster-b>
```

The secrets referenced here must currently contain the kubeconfig for a the Kubenetes cluster available at the provided URL as the value of secret key 'kubeconfig'.
This allows the central Strimzi operator to authenticate with multiple Kubernetes clusters.
However, this approach introduces two key challenges:

- Security Risk – The Kubeconfig file stored in these secrets grants broad access to the entire remote cluster. 
If a user provides a full Kubeconfig, it could unintentionally expose all cluster resources, creating a potential security vulnerability.

> - Short-Lived Tokens – The tokens within the Kubeconfig files are short-lived (typically expiring within 48 hours), requiring frequent renewal and adding complexity to long-term automation.

Reviewer 2: I do not think this makes any sense. Kubeconfig file is a format for storing credentials and connection information to one or more Kubernetes clusters. It does not say anything about your privileges, about the authentication mechanism, about the expiration of the tokens and you can limit the permissions it has in the same way as you would do for a service account. So IMHO the justification of why kubeconfig cannot be used makes any sense.

I'm not sure if you can replace the kubeconfig in 100% of cases with a _service account_ and I do not think the service account is designed to be copied around. But I bet you can put your long-lived token into kubeconfig file.

Also, keep in mind that this is not only about the token but also about things such as trusted TLS certificates / user TLS certificates / usersnames / password etc. which are often required for outside access etc.

Author: You're absolutely right that Kubeconfig itself is just a format, and its security depends on the credentials it contains. Our concern is not with Kubeconfig itself but with ensuring that the credentials inside it have minimal, necessary permissions.

In our prototype, the Kubeconfig stored in the secret should not include broad user credentials with cluster-wide access. Instead, we generate a Kubeconfig that contains a ServiceAccount token with only the required permissions for the Strimzi operator. This is achieved by:

- Creating a dedicated ServiceAccount in the remote cluster.
- Binding it to a Role with minimal privileges (e.g., managing only Kafka-related resources).
- Generating a long-lived token for this ServiceAccount and embedding it in the **`Kubeconfig`**. (Hope this part is clear)

This ensures that the Strimzi operator can securely authenticate with remote clusters while preventing accidental overexposure of cluster-wide resources.

Reviewer 2: I don't think its as simple. As I said, different users use different configurations. My Kube clusters for example use mTLS for remote communication. So yu have to keep in mind that while guiding the users to do secure things is good, you also need flexibility to make sure people can use it.

Author: That's a great point. Our goal is to provide a secure default while allowing flexibility for different authentication mechanisms, including mTLS. The proposed approach focuses on ensuring that, when using token-based authentication, the credentials in Kubeconfig are scoped with minimal privileges. However, we recognize that different users may have different security requirements, and we’ll make sure the design allows for configurable authentication options, including mTLS.

Reviewer 2: The point I'm trying to make is that it is primarily the user's responsibility to secure the credentials. All we need is to document it, likely provide some RBAC YAMLs with the required rights etc. But at the software level, Kubeconfig is what the operator should consume, and what is inside should not really matter that much to the operator.

Reviewer 4: @scholzj You're absolutely right; while the documentation should provide examples to help folks to get going quickly, the responsibility for securely administering their Kubernetes infrastructure should fall to the user. Potentially there could be tooling to simplify the initial configuration, but that would be a "nice-to-have".

Thank you for bringing the perspective we needed here :)

Author: updated proposal to clarify this

Unlike MirrorMaker 2 (MM2), a stretch cluster provides strong data durability through synchronous replication and enables fast disaster recovery with automated client failover

- **Migration Flexibility**: A stretch Kafka cluster enables seamless migration, whether it's moving the entire cluster across Kubernetes environments or cloud providers without downtime, or relocating individual Kafka nodes as needed. 
This flexibility helps with maintenance, scaling, and workload transitions between environments.

- **Resource Optimization**: Efficiently utilizing resources across multiple clusters, which can be advantageous in environments with varying cluster capacities or during scaling operations.

## Proposal

This proposal seeks to enhance the Strimzi Kafka operator to support stretch Kafka clusters, distributing broker, controller and combined Kafka nodes across multiple Kubernetes clusters.
The intent is to focus on high-availability of the Kafka data plane.

> The proposal outlines high-level topology and design concepts for such deployments, with a plan to incrementally include finer design and implementation details for relevant aspects.

Author: The statement about "incrementally including finer design and implementation details" was originally added when the proposal was first drafted and still in its early stages. Since then, the proposal has evolved significantly, and most of the key design and implementation aspects are now already included.

Given the current completeness of the proposal, we agree that this statement is no longer accurate or necessary. We will remove or update it to better reflect the current state of the proposal.

Thanks again for catching this

As part of this mode:
The custom actions taken as a result of this environment variable include:
1. The value of `STRIMZI_NETWORK_POLICY_GENERATION` is always treated as **`false`**, regardless of the value provided by the user or the default (true).
This is to ensure that network traffic between Kubernetes clusters is not blocked due to network policies.
The Cluster Operator will log a warning to inform the user that any value set for this variable is being ignored in stretch mode.
2. The `StrimziPodSet` resources created in a remote cluster will include an annotation `strimzi.io/remote-podset: true`.
This annotation will allow the remote cluster operator to reconcile the `StrimziPodSet` without a `Kafka` and `KafkaNodePool` CR being present in the same cluster.

The `STRIMZI_STRETCH_CLUSTER_NETWORK` environment variable defines the cloud-native networking technology used for cross-cluster communication in a stretch Kafka deployment.
Specifying this value in the Cluster Operator deployment is essential because different networking technologies may require different handling or integration steps by the operator.

> For example, when using Submariner, the operator needs to create additional Kubernetes resources such as ServiceExport objects to enable seamless service discovery and communication across clusters.

Author: Yes, that's true. But in my opinion, we don't need to worry about this, as ServiceExport is only required for Submariner-based stretch clusters. For the initial implementation, we're proposing a Cilium-based stretch cluster since it's a graduated CNCF project and offers excellent performance, with low latency.

> For example, when using Submariner, the operator needs to create additional Kubernetes resources such as ServiceExport objects to enable seamless service discovery and communication across clusters.

Reviewer 3: Ok but by reading the proposal my understanding was that we were going to support both and defaulting to Cilium.
Having Submariner support means additional RBAC and also additional code into the operator to create the needed additional resources. If you are mentioning the Submariner usage in the proposal I would expect to have such a code already (in the POC?) so RBAC should be mentioned.
If, instead, the proposal is about supporting Cilium only (which could make sense anyway), then we should make it clear and set as a "no goal" for this proposal, the support for other networking projects that could have different requirements. 
Or we could think about a pluggable way, thinking about an interface between operator and network specific implementation that could be also brought by the user (I am just thinking aloud about this).

> For example, when using Submariner, the operator needs to create additional Kubernetes resources such as ServiceExport objects to enable seamless service discovery and communication across clusters.

Author: Our POC supports both Submariner and Cilium, but the current plan is to proceed with just one - Cilium, for the initial implementation. I'll revisit the entire proposal and clarify any confusing elements.

> For example, when using Submariner, the operator needs to create additional Kubernetes resources such as ServiceExport objects to enable seamless service discovery and communication across clusters.

Reviewer 4: I've moved Submariner to be a "Rejected alternative" for clarity as suggested and given the reasons. I've also tried to clarify the extension point provided by the environment variable.

> For example, when using Submariner, the operator needs to create additional Kubernetes resources such as ServiceExport objects to enable seamless service discovery and communication across clusters.

Author: I still think it's reasonable to require the user to provide a value for the cross-cluster network type, even if we currently support only one option (e.g., Cilium). Introducing this environment variable now helps us ensure that the value is consistently used throughout the Strimzi codebase from the start. This not only avoids assumptions but also sets a clear extension point, making it much easier to add support for additional technologies in the future without needing code changes or introducing backward compatibility issues later on.

This approach also makes the implementation extensible and future-proof, as additional technologies can be supported by adding their respective handling logic to the operator based on this variable.

If the `STRIMZI_STRETCH_CLUSTER_NETWORK` environment variable is not explicitly defined, the Cluster Operator will default to "cilium", as it does not require any operator-driven configuration for cross-cluster service discovery.
A warning will be logged to inform users of this default behavior.
The warning will look something like:

```
STRIMZI_STRETCH_CLUSTER_NETWORK is not defined. Defaulting to 'cilium'.
```

This approach ensures that stretch mode remains functional even if the variable is omitted.
However, if the user has not set up any cross-cluster networking technology (such as Cilium, Submariner, or others) in the underlying Kubernetes clusters, the stretch cluster will not function correctly regardless of the operator configuration.
It is the user’s responsibility to ensure that appropriate cross-cluster networking is in place for pod-to-pod communication across clusters.

##### Remote cluster operator configuration

> To optimize resource usage in the remote cluster, the operator can optionally be configured to reconcile only `StrimziPodSet` resources by setting the existing environment variable:

Author: Thanks for raising this important point!
If `STRIMZI_POD_SET_RECONCILIATION_ONLY` is not set, the operator in the remote cluster will attempt to reconcile any Kafka, KafkaNodePool, KafkaConnect resources that might be created there, which could cause confusion

In a stretch cluster deployment, Kafka/KafkaNodePool CRs should only exist in the central cluster. However, to prevent mistakes or accidents in the remote cluster, it is strongly recommended to set STRIMZI_POD_SET_RECONCILIATION_ONLY=true when deploying the operator to remote clusters.

I'll update the proposal to make this clear and more opinionated, recommending setting this variable explicitly to safeguard against accidental CR creation in remote clusters.

> To optimize resource usage in the remote cluster, the operator can optionally be configured to reconcile only `StrimziPodSet` resources by setting the existing environment variable:

Author: @katheris I have updated the proposal to make this clearer. Could you please take a look?

> To optimize resource usage in the remote cluster, the operator can optionally be configured to reconcile only `StrimziPodSet` resources by setting the existing environment variable:

Reviewer 3: Setting this env var to true is important and not so optional. Without it, the operator would start to reconcile Kafka, KafkaNodePool and create a new cluster locally with the same name of the strech one, also the services, the PVCs, ... something that the central cluster is doing at the same time (because of the strech cluster) etc etc ... it could be a mess. At this point I would say that if you want to deploy a strech cluster and you specify to do so through the operator, this env var should be ignored and thread as true on the remote cluster instead?

> To optimize resource usage in the remote cluster, the operator can optionally be configured to reconcile only `StrimziPodSet` resources by setting the existing environment variable:

Author: @ppatierno 

Thank you for the feedback and the suggestion to programmatically enforce `STRIMZI_POD_SET_RECONCILIATION_ONLY=true` on remote clusters.

I'm thinking aloud here 🤔🤔🤔, I would like to highlight that this is a bit more complex compared to how we handle, for example, the `STRIMZI_NETWORK_POLICY_GENERATION` variable.

In the case of `STRIMZI_NETWORK_POLICY_GENERATION`, the Operator can detect that the Kafka cluster spans multiple Kubernetes clusters  (stretch deployment) and internally force network policy generation to be disabled. This is possible because network policies are only created in the cluster where the Kafka and KafkaNodePool CRs are deployed (The central cluster).

However, for `STRIMZI_POD_SET_RECONCILIATION_ONLY`, the situation is different.
This variable needs to be set individually for the Cluster Operator instances running in each remote cluster, because it is the Operator running in the remote cluster that should reconcile only `StrimziPodSet` resources.

In the current stretch cluster design, the Cluster Operator running in the central cluster reconciles the Kafka and KafkaNodePool resources and creates the necessary resources (SPS, ConfigMaps, Secrets, Services, etc) in the remote clusters by interacting with their Kubernetes APIs to bring Kafka Pods in the respective clusters. However, the central Operator does not manage or patch the Operator deployments running in the remote clusters.
There is no Operator-to-Operator communication, and modifying the remote Operator's environment variables or deployments dynamically is currently out of scope for the current stretch implementation.

Technically, it might be possible to patch the remote Operator deployments from the central Operator in the future, but this would require a significant amount of additional work and complexity to implement cleanly and securely.

That is why I originally suggested documenting this clearly and strongly recommending users to set `STRIMZI_POD_SET_RECONCILIATION_ONLY=true` when deploying the Operator to remote clusters.

Of course, if we feel that documentation alone is not sufficient and that automatic enforcement is absolutely necessary, I’m happy to work with the community to explore clean and standard approaches to solve this

> To optimize resource usage in the remote cluster, the operator can optionally be configured to reconcile only `StrimziPodSet` resources by setting the existing environment variable:

Reviewer 4: @katheris I've cleaned up the language here to clarify that we should document setting `STRIMZI_POD_SET_RECONCILIATION_ONLY` to `true` is mandatory on all remote Kubernetes clusters.

@ppatierno Are you suggesting we programmatically lock this configuration based on the configuration of the operator on the central Kubernetes cluster? I'm not sure we should have a situation where one operator is directly modifying the configuration of an operator in another Kubernetes; That seems like an anti-pattern.

> To optimize resource usage in the remote cluster, the operator can optionally be configured to reconcile only `StrimziPodSet` resources by setting the existing environment variable:

Author: Thanks for the latest commit, @mark-VIII. The changes look clear.
@ppatierno @katheris, could you please review?

  namespace: strimzi
type: Opaque
data:
  kubeconfig: <base64-encoded-kubeconfig>
```

When the `STRIMZI_REMOTE_KUBE_CONFIG` environment variable is set, the Cluster Operator will recognize that it needs to deploy a stretch Kafka cluster. 

The custom actions taken as a result of this environment variable include:
1. The value of `STRIMZI_NETWORK_POLICY_GENERATION` is always treated as **`false`**, regardless of the value provided by the user or the default (true).
This is to ensure that network traffic between Kubernetes clusters is not blocked due to network policies.
The Cluster Operator will log a warning to inform the user that any value set for this variable is being ignored in stretch mode.
2. The `StrimziPodSet` resources created in a remote cluster will include an annotation `strimzi.io/remote-podset: true`.
This annotation will allow the remote cluster operator to reconcile the `StrimziPodSet` without a `Kafka` and `KafkaNodePool` CR being present in the same cluster.

The `STRIMZI_STRETCH_CLUSTER_NETWORK` environment variable specifies the cloud-native networking technology used for cross-cluster communication in a stretch Kafka deployment.

> Allowing this additional configuration for the Cluster Operator deployment provides an extension point for additional CNCF networking projects to be used.

Reviewer 3: This would not be enough anyway. Taking Submariner as an example, it would need code changes within the operator and not just setting a different value for this env var.
Also, if the Cilium usage doesn't need any code changes in the operator, what is the operator doing with this env var in the current POC? I haven't looked at it in detail yet.

Author: The STRIMZI_STRETCH_CLUSTER_NETWORK environment variable is intended to identify which cloud-native networking technology is being used for cross-cluster communication.

The main motivation for introducing this environment variable is to allow the Cluster Operator to perform any networking-specific actions that might be needed to support a stretch Kafka deployment, depending on the underlying technology.

As you have rightly pointed out, when Submariner is used, the Strimzi Operator needs to create additional resources like ServiceExport objects to enable cross-cluster service discovery. In contrast, for Cilium-based environments, no such additional steps are currently necessary --> meaning that for Cilium, the environment variable does not actively affect the operator’s behavior at the moment.

However, we are proposing the inclusion of this environment variable with a future-looking perspective. As we integrate support for more cloud-native networking technologies over time, the operator might need to perform specific actions for those as well. Having this configuration available from the start would allow us to extend support more easily without major structural changes later. That said, we fully acknowledge that today, for technologies like Cilium, this setting might not be strictly needed.

For an example of how Submariner-based stretch clusters make use of this environment variable within the operator, please [see](https://github.com/aswinayyolath/strimzi-kafka-operator/commit/c4c7b49fc49e68e242f917bd4acfb0a807af8377#diff-ac9b17492d6e343d1a402e70ab4460b2f56e6d3a66600f707cb354523384c651R28-R176)

Reviewer 3: > However, we are proposing the inclusion of this environment variable with a future-looking perspective

But the future could be:

* Adding support for another network tech which acts like Cilium so without the need for additional resources to be created (and so no changes in the operator). In this case the env var will be still "useless" (not influencing the operator).
* Adding support for another network tech, like Submariner (or like it), which needs additional intervention by the operator which would need some code changes to create resources. Only in this case the env var will start to be useful to distinguish different code paths (i.e. Cilium vs Submariner). But is it worth having the env var right now which looks to be totally unused for what you are proposing with Cilium right now?

I would like to know even what other maintainers think about that @strimzi/maintainers ?

Author: Thank you for the feedback - I understand the concern about introducing unused configuration.

We are happy to drop the environment variable from the initial implementation if the consensus is that it’s premature.

Alternatively, another option could be to introduce the env var only at the point when support for a networking technology that requires operator-side actions (like Submariner) is formally added.

Our primary motivation was to avoid a disruptive change later (introducing new env var post-release usually requires additional versioning and documentation overhead). However, this is considered too speculative at this point, we are fully fine with deferring the addition.

We can move forward with a minimal Cilium focused impementation without the env var, and revisit it when necessary.

Thanks again for the careful review! ❤️

Reviewer 4: I have added another commit to clarify the requirement for a Multicluster Services API implementation of which Cilium is the first supported. I think this nullifies the need for the `STRIMZI_STRETCH_CLUSTER_NETWORK` environment variable so I have removed that.

@aswinayyolath I believe this update will mean we can resolve this conversation thread.

Reviewer 2: @mark-VIII I think this is quite significantly confusing building the support based on an API and on a particular technology/project. If you say that it is based on MCS API, then it should support any MCS API implementation that is conformant with it. So saying that you _build it on MCS API with the Cilium as a first supported implementation_ suggests that there is some problem somewhere. For example that the API is not well designed such as the Ingress API for example.

Reviewer 4: @scholzj I see what you are saying and you make a good point. 

I have updated the proposal to address this and removed references to specific implementations.

# Stretch Kafka cluster

The Strimzi Kafka Operator currently manages Kafka clusters within a single Kubernetes cluster.
This proposal aims to extend its support to stretch Kafka clusters, where the brokers and controllers of a single Kafka cluster are distributed across multiple Kubernetes clusters.

## Current situation

At present, the availability of Strimzi-managed Kafka clusters is constrained by the availability of the underlying Kubernetes cluster.
If a Kubernetes cluster experiences an outage, the entire Kafka cluster becomes unavailable, disrupting all connected Kafka clients.

While it is possible to enhance availability by running Kubernetes across multiple availability zones and configuring Strimzi with affinity rules, tolerations, or topology spread constraints, this approach is limited to a single Kubernetes control plane. 

> Additionally, if not self-hosted, deploying across multiple availability zones typically incurs higher cloud hosting costs. 

Author: I've revised the section to clarify the limitations of running across availability zones within a single Kubernetes cluster. The updated text now emphasizes the risk of relying on a single Kubernetes control plane and contrasts that with the improved fault tolerance of the stretch cluster, which uses independent clusters with separate control planes.

> Additionally, if not self-hosted, deploying across multiple availability zones typically incurs higher cloud hosting costs. 

Reviewer 2: I do not think that is really sufficient. If you think that a single HA control-plane is not sufficient on the Kubernetes level, why do you think it is sufficient on the Kafka level? Kubernetes control-plane is mature, hardened, and stable thing, battle-tested by many users. Compare it to KRaft control-plane which is brand new, has unclear number of users, it is still under development with many limitations. Which one would you want to rely on a single HA control plane with? I definitely with Kubernetes!

So why is it so important to have multiple independent Kubernetes clusters and stretch across them a single Kafka cluster?

> Additionally, if not self-hosted, deploying across multiple availability zones typically incurs higher cloud hosting costs. 

Author: You’re right that K8s control planes are mature and generally resilient — we’re not arguing that a K8s control plane is inherently fragile. Rather, our goal is to acknowledge a broader operational reality: K8s clusters are typically isolated failure domains.

While a single multi-AZ K8s cluster with an HA control plane may offer HA within that cluster, there are operational and organizational reasons why teams prefer to run independent clusters per zone, region, or DC - for fault isolation, compliance, upgrades, or team autonomy.

Strtch Kafka supports that model by enabling a Kafka deployment that spans these independently managed clusters, without requiring users to consolidate everything into one K8s control plane.

Also, the rationale is not that KRaft is more resilient than K8s --> it’s that Kafka’s quorum-based architecture can tolerate the failure of a subset of its nodes (including controller nodes), as long as a quorum remains. This allows Kafka to survive the complete loss of one K8s cluster, something a single cluster deployment can’t do — no matter how resilient its control plane is.

We’ll clarify that our focus is on failure isolation at the infrastructure level, not making a blanket claim about control plane maturity.

> Additionally, if not self-hosted, deploying across multiple availability zones typically incurs higher cloud hosting costs. 

Reviewer 2: > Also, the rationale is not that KRaft is more resilient than K8s --> it’s that Kafka’s quorum-based architecture can tolerate the failure of a subset of its nodes (including controller nodes), as long as a quorum remains. This allows Kafka to survive the complete loss of one K8s cluster, something a single cluster deployment can’t do — no matter how resilient its control plane is.

Of course a single cluster deployment can do it. When you have stretched Kubernetes cluster over 3 AZs with Kafka cluster over all 3 AZs, it both survives crash one one AZ. Kubernetes as well as KRaft. The architecture of Kubernetes control plane and Kafka control plane are not fundamentally different.

----

I can understand why you might want to run a separate Kubernetes cluster per AZ. It does likely give you a slightly better availability on the Kubernetes level over a single Kubernetes cluster stretched over multiple AZs.

What does not make that much sense to me is that you invest a huge amount of effort (both operational, but also compute resources as you now need to run 3 separate HA Kubernetes control planes instead of one) to slightly improve the availability of the Kubernetes level.

But then just negate most of it by not doing the same on the Kafka level, where you are still dependent on a single control plane instead of having a separate Kafka cluster per-Kubernetes cluster, and mirror the availability model you use for Kubernetes.

The architecture should have some symmetry on all its layers. They cannot be completely disconnected from each other.

> Additionally, if not self-hosted, deploying across multiple availability zones typically incurs higher cloud hosting costs. 

Author: I've already removed 

```
Additionally, if not self-hosted, deploying across multiple availability zones typically incurs higher cloud hosting costs. 
```
Marking this thread as resolved

A stretch Kafka cluster offers an alternative approach by distributing Kafka brokers and controllers across multiple Kubernetes clusters, reducing dependency on a single cluster's control plane while improving resilience.

## Motivation

A stretch Kafka cluster allows Kafka nodes to be distributed across multiple Kubernetes clusters. This approach also facilitates many valuable use cases, such as:

> - **High Availability**: Distributing Kafka brokers across multiple Kubernetes clusters significantly enhances resilience by enabling the system to tolerate the outage of an entire Kubernetes cluster without disrupting service to clients. 

Author: I’ve updated the High Availability section to clearly address how a stretch cluster improves resilience compared to a single stretched Kubernetes cluster. This now highlights the independence of control planes across clusters, ensuring that the failure of one cluster doesn't impact Kafka brokers in others.

Unlike MirrorMaker 2 (MM2), a stretch cluster provides strong data durability through synchronous replication and enables fast disaster recovery with automated client failover

> - **Migration Flexibility**: A stretch Kafka cluster enables seamless migration, whether it's moving the entire cluster across Kubernetes environments or cloud providers without downtime, or relocating individual Kafka nodes as needed. 

Reviewer 2: You list this here, but the proposal is completely missing how it would actually work.

This flexibility helps with maintenance, scaling, and workload transitions between environments.

> - **Resource Optimization**: Efficiently utilizing resources across multiple clusters, which can be advantageous in environments with varying cluster capacities or during scaling operations.

Author: I've expanded the `Resource Optimization` section to provide clearer justification. I completely understand your concern about the need for symmetry when HA  is the primary goal.
Our intention is not to suggest extreme asymmetry or uncontrolled placement, but rather to highlight that moderate, controlled flexibility can be beneficial in real-world deployments where clusters may naturally differ in available resources. A few additional points we would like to share:

1. Stretch clusters allow us to expand only those clusters that have spare capacity, rather than being forced to overprovision uniformly across all clusters. This helps avoid unnecessary infrastructure costs.
2. In many hybrid or multi-cloud setups, it’s common to have clusters with different sizes or resource quotas. Stretch clusters provide a way to take advantage of these differences while still maintaining quorum and rack-awareness.
3. During maintenance or upgrades in one cluster, the healthy clusters can continue to serve clients without causing downtime, as long as quorum is preserved.
We believe this offers an operational optimization as well, beyond just raw resource usage.

So to summarize , We fully agree that availability remains the primary concern, and that quorum and placement rules must be carefully managed. Our goal here is to point out that stretch deployments can provide some operational and scaling flexibility, without compromising the resilience objectives of the system.

> - **Resource Optimization**: Efficiently utilizing resources across multiple clusters, which can be advantageous in environments with varying cluster capacities or during scaling operations.

Reviewer 2: PS: I do not want to just be nitpicky here. But this will be used as the core for the documentation, and it will be read by our users deciding how to (not) use this feature. So you have to be able to explain the situations where it makes sense and explain what it actually brings.

> - **Resource Optimization**: Efficiently utilizing resources across multiple clusters, which can be advantageous in environments with varying cluster capacities or during scaling operations.

Author: Thank you very much for the thoughtful feedback,  I truly apreciate your concern about how this section could be interpreted, especially given its importance for future documentation and user guidance.

You're absolutely right that when HA is the primary goal, stretch Kafka clusters generally require symmetric provisioning across all sites to maintain quorum and avoid single points of failure. We are not suggesting that users should run production grade HA deployments with imbalanced or under-provisioned clusters, that would indeed be a risky and unsupported pattern.

What we’re trying to highlight here is a more operational and transitional flexibility that stretch clusters can offer, without compromising the core HA guarantees. For eg:

If one site temporarily has more available capacity, it can be leveraged for scaling or rebalancing workloads during peak times or phased expansions  as long as quorum and rack-awareness are still preserved. During rolling upgrades or maintenance windows, the cluster can remain available through brokers in the other site(s), avoiding downtime without needing full overprovisioning everywhere. In hybrid or multi-cloud setups, where budget, approvals, or physical infrastructure availability vary, stretch can act as a bridge for gradual, controlled rollout not as a primary strategy for significant cost reduction, but as a practical operational benefit. We’ve updated the **Resource Optimization** section to reflect this more clearly and to avoid any impression that asymmetric stretch clusters are a recommended deployment model for HA. The emphasis is now on how stretch supports flexibility during operations, rather than being a core mechanism for reducing resource usage.

> - **Resource Optimization**: Efficiently utilizing resources across multiple clusters, which can be advantageous in environments with varying cluster capacities or during scaling operations.

Reviewer 2: Sorry, I do not understand how you mean it. If you want to maintain the high availability, you need to have the capacity on all sites as you need the data to be replicated. You could I guess - in theory - try to do something like having all the partition leaders in one site, but how much difference does it make? Few percent of the CPU? But you also need to be able to switch quickly to the other site if you want to be highly available, so the available resources cannot be too skewed. So where do you get the difference?

> - **Resource Optimization**: Efficiently utilizing resources across multiple clusters, which can be advantageous in environments with varying cluster capacities or during scaling operations.

Author: Thanks again for the continued feedback, it's much appreciated. I understand the concerns, and you're right  maintaining HA in stretch Kafka clusters does require symmetric provisioning across sites to uphold quorum and failover guarantees.

Our intention with the `Resource Optimization` section was to highlight some operational and transitional flexibility in specific scenarios (e.g., phased provisioning, rolling maintenance, or short-term scaling in one site with spare capacity). However, since this isn't a core objective of the proposal and could easily be misinterpreted as advocating asymmetric provisioning for HA, I've removed the section to avoid any confusion and keep the focus clear on resilience and availability.

> - **Resource Optimization**: Efficiently utilizing resources across multiple clusters, which can be advantageous in environments with varying cluster capacities or during scaling operations.

Reviewer 2: If you see that as a separate usecase where someone does this independently on the HA aspect, maybe it makes sense for someone. So I do not necessarily want to throw it out if it makes sense to you. But it seems to be the use-cases might need at least some more clarification.

## Proposal

This proposal seeks to enhance the Strimzi Kafka operator to support stretch Kafka clusters, distributing broker, controller and combined Kafka nodes across multiple Kubernetes clusters.

> The intent is to focus on high-availability of the Kafka data plane.

Author: I completely see how the wording was confusing here.

The intention behind the original line "The intent is to focus on high-availability of the Kafka data plane." was to emphasize that our primary goal is to ensure that Kafka client operations (e.g., producing and consuming messages) continue to function even if one of the Kubernetes clusters -> including the central cluster where the control plane resides, experiences a failure.

In this context, when we said "data plane," we were referring to Kafka's ability to serve client traffic, not just the presence of broker-only nodes. I now realize that this terminology can easily be misinterpreted to mean a focus solely on broker pods, excluding controller or mixed nodes, which is not our intent.

To avoid this ambiguity, I've rephrased the sentence in the latest update to make it clear that the proposal's goal is to provide client-level high availability, even in the event of central or remote cluster failure. Let me know if the new phrasing better aligns with how you'd expect this to be communicated.

> The intent is to focus on high-availability of the Kafka data plane.

Author: I’ve updated the proposal to clearly state that both broker and controller Pods will be distributed across all the involved Kubernetes clusters. This ensures that the Kafka data plane and control plane are both stretched, allowing the cluster to continue serving client operations even in the event of a full failure of any single cluster, including the one hosting Kafka, KNP CRs.

To further support this, I’ve already implemented and demonstrated this setup in action. For reference, here’s a short recording of the demo that shows the brokers and controllers spanning across multiple K8s clusters: 
 https://www.youtube.com/watch?v=NEPgtXD6voA

The proposal outlines the high-level topology, design concepts, and detailed implementation considerations for such deployments.

### Limitations and Considerations
While a stretch Kafka cluster offers several advantages, it also introduces some challenges and considerations:

- **Increased Network Complexity and Costs**: The communication between brokers and controllers across clusters relies on network connectivity, which can be less reliable and more costly than intra-cluster communication.
This necessitates careful consideration of network architecture and associated costs.

- **Latency Requirements**: Stretch Kafka clusters are best suited for environments with low-latency and high-bandwidth network connections between the Kubernetes clusters.
High latency can adversely affect the performance and synchronization of Kafka nodes, potentially leading to delays or errors in replication and client communication.
Defining the minimal acceptable latency between clusters is crucial to ensure optimal performance.

### Prerequisites

- **KRaft**: This proposal focuses exclusively on enabling stretch deployments for KRaft-based Kafka clusters.

> Strimzi will not support stretch clusters for Zookeeper-based deployments.

Author: Agreed, I will remove

It is not feasible to test stretch cluster functionality with every MCS implementation.
Therefore, a single implementation will be selected for test purposes that is most closely aligned with Strimzi test infrastructure.

### High-level architecture

#### Topology of a stretch cluster

![Stretch cluster topology](./images/095-stretch-cluster-topology.png)

The diagram illustrates a topology comprising of three Kubernetes clusters.

One of these clusters is designated as the "Central cluster", while any additional clusters are considered "remote".
The central cluster acts as the control plane where a user will create all the custom resources for the Kafka cluster - Kafka, KafkaNodePool, KafkaUser, KafkaTopic etc.

A Kafka node pool definition can be configured to specify a Kubernetes cluster (central cluster or one of the remote clusters) as the deployment target.
The operator on the central cluster is responsible for creating all necessary resources for the node pool on the target Kubernetes cluster. 

### Low-level design and prototype

The following sections will describe the key aspects of the proposal.

#### Step 1: User configuration of the chosen MCS-API implementation

> When configuring Multicluster Services the user must specify a unique identifier for each Kubernetes cluster.

Author: It is the user’s responsibility to configure cloud-native multicluster networking tools such as Submariner, Cilium ClusterMesh, etc., when setting up a Stretch Kafka cluster. Each of these tools requires the user to explicitly assign a `unique identifier` to each K8s cluster as part of their installation process. The Strimzi operator does not handle installation or configuration of these tools directly, it only expects that this unique identifier is already available and can be referenced when generating Kafka configurations or targeting remote clusters.

For eg:

Cilium ClusterMesh users provide a unique cluster ID via `--set cluster.name=<name> and --set cluster.id=<id>` flags during installation using cilium install. Once Cilium is deployed, they also establish inter-cluster connectivity via cilium clustermesh connect.

Submariner users specify the cluster identity using the `--clusterid` flag while running subctl join. This is done when joining each cluster to the Submariner broker.

Strimzi remains implementation-agnostic and simply consumes the configured cluster identifiers as part of the Stretch Kafka deployment flow. The installation steps for Submariner and Cilium are documented [here](https://aswinayyolath.github.io/stretch-kafka-docs/Installing-Submariner-on-OpenShift/) and [here](https://aswinayyolath.github.io/stretch-kafka-docs/Setting-up-cilium/). We have deliberately excluded these details from the proposal to keep it concise and focused. Including installation instructions would make the proposal lengthy and harder to follow, and in any case, Strimzi should not be responsible for the installation or management of external multicluster networking solutions.

> When configuring Multicluster Services the user must specify a unique identifier for each Kubernetes cluster.

Reviewer 4: The approach proposed by the Kubernetes multicluster SIG is to support the About API: https://multicluster.sigs.k8s.io/concepts/about-api/

The demo video they link to (https://multicluster.sigs.k8s.io/#getting-started) uses this API and is based upon EKS with AWS Cloud Map and the AWS MCS controller. However, as @aswinayyolath describes, some MCS implementations might do some of this configuration under the covers or use a different mechanism as they haven't get aligned with the API specs.

> When configuring Multicluster Services the user must specify a unique identifier for each Kubernetes cluster.

Reviewer 2: So, you expect that to be the same as the `cluster-id` / `alias` in the annotation and in the configuration of the Kubeconfigs?

> When configuring Multicluster Services the user must specify a unique identifier for each Kubernetes cluster.

Author: Yes, that's correct — the same cluster identifier configured during the installation of the multicluster networking layer (e.g., `--clusterid` in Submariner or `--set cluster.id` in Cilium) is expected to be reused as the `stretch-cluster-alias` in the KafkaNodePool annotation and also as the key in the `STRIMZI_REMOTE_KUBE_CONFIG` environment variable. This ensures consistency when generating Kafka configurations and targeting remote clusters.

> When configuring Multicluster Services the user must specify a unique identifier for each Kubernetes cluster.

Reviewer 2: Ok, thanks for the clarification.

  value: |
      cluster-id-a.url=<cluster-a URL>
      cluster-id-a.secret=<secret-name-cluster-a>
      cluster-id-b.url=<cluster-b URL>
      cluster-id-b.secret=<secret-name-cluster-b>
```

The secrets referenced here must contain the kubeconfig for the Kubernetes cluster available at the provided URL as the value of secret key 'kubeconfig'.
This allows the central Strimzi operator to authenticate with multiple Kubernetes clusters.

**Example Secret**

Below is an example Kubernetes Secret containing a kubeconfig for a remote cluster

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: secret-name-cluster-a
  namespace: strimzi
type: Opaque
data:
  kubeconfig: <base64-encoded-kubeconfig>
```

When the `STRIMZI_REMOTE_KUBE_CONFIG` environment variable is set, the Cluster Operator will recognize that it needs to deploy a stretch Kafka cluster. 

The custom actions taken as a result of this environment variable include:
1. The value of `STRIMZI_NETWORK_POLICY_GENERATION` is always treated as **`false`**, regardless of the value provided by the user or the default (true).
This is to ensure that network traffic between Kubernetes clusters is not blocked due to network policies.

> The Cluster Operator will log a warning to inform the user that any value set for this variable is being ignored in stretch mode.

Author: In our early stretch cluster experiments using an `Ingress-based` manual setup, we encountered issues with the default Strimzi netpol blocking cross-cluster traffic, which required us to delete the policy manually. Based on that experience, we initially assumed the same would apply to other stretch configurations and carried that logic into the proposal. However, after further testing with **Submariner** and **Cilium**, we validated that this assumption does not hold in MCS-based solutions. In both cases, we tested the stretch setup without disabling `STRIMZI_NETWORK_POLICY_GENERATION`, and confirmed that the default Netpol created by Strimzi in the central cluster does not block cross-cluster traffic. This is because technologies like Submariner and Cilium create a unified network layer that transparently routes traffic between clusters without triggering the restrictions of intra-cluster Netpols. As such, disabling network policy generation is unnecessary, and we've [updated the proposal ](https://github.com/strimzi/proposals/pull/129/commits/80d7e2cb5c448071ac81675216325155a97d3118) accordingly. This was a simple oversight carried forward from an earlier test setup, and we appreciate the review feedback that helped clarify this point.

2. The `StrimziPodSet` resources created in a remote cluster will include an annotation `strimzi.io/remote-podset: true`.
This annotation will allow the remote cluster operator to reconcile the `StrimziPodSet` without a `Kafka` and `KafkaNodePool` CR being present in the same cluster.

##### Remote cluster operator configuration

> When deploying the operator to remote clusters, the operator must be configured to reconcile only StrimziPodSet resources by setting the existing environment variable:

Reviewer 5: There was a separate thread discussing this previously https://github.com/strimzi/proposals/pull/129#discussion_r2046696257. For some reason it was marked resolved.

The main discussion on that thread was around causing confusion if the user creates a `Kafka` resource on the remote cluster with the same name as the one on the central cluster.

I feel like this is a tricky one that I could argue either way.

If the env var is documented as mandatory:

- There is no way to programmatically enforce it, since as discussed on the previous thread the remote cluster operator doesn't know it's on a remote cluster. So I think even with good documentation we need to accept that people will deploy without the env var being set.

But if we say it isn't mandatory there are a bunch of extra things we probably need to consider and implement:

- What happens if users deploy Kafka resources with the same name as the one in the central cluster? Do we need an extra check that looks for existing StrimziPodSet resources and if they exist check they don't have the `remote-podset` annotation?
- For Connect and MM2 what happens if the user creates the clusters using the same named topics on different Kubernetes clusters? Do we rely on the user to properly configure their topics, or should the operator be adding the stretch cluster alias to the topic names to remove that issue?

I can see the appeal of documenting it as mandatory in the initial implementation, but then later making it optional. However I still think that what is missing from this proposal is how a user is expected to make use of a stretch cluster within the wider landscape of Strimzi components. Once we have that defined I think it will help inform what we should do here

> When deploying the operator to remote clusters, the operator must be configured to reconcile only StrimziPodSet resources by setting the existing environment variable:

Author: I have opened that thread again!

> When deploying the operator to remote clusters, the operator must be configured to reconcile only StrimziPodSet resources by setting the existing environment variable:

Reviewer 2: I think it cannot be mandatory as you cannot enforce it, but also because it would mean you cannot use any other Strimzi operands along side of it. What is worth the stretch Kafka cluster if you cannot run your Connect or Bridge next to it? (I also raised a separate comment for this as there needs to be some guidance on how to run those even if we currently do not propose stretch Connect clusters for example).

##### Kafka CR

There is no change required to the Kafka custom resource (Kafka CR).
The Kafka resource remains exactly as defined in the existing Strimzi API, and users can continue to define their cluster configurations in the standard way.
Remote cluster access credentials are configured through environment variables in the Cluster Operator deployment.
This simplifies the Kafka CR and keeps global operational concerns decoupled from workload specifications.
Cluster-specific configuration, such as the target remote cluster for each node pool, is specified in the KafkaNodePool CR via annotations.

##### KafkaNodePool CR

The `KafkaNodePool` CR remains in the central cluster but must include a new annotation to indicate the target Kubernetes cluster where the corresponding Kafka node resources should be deployed.
Based on this annotation, the Cluster Operator will create the associated resources — including StrimziPodSet, Services, Secrets, ConfigMaps, and others — in the specified remote cluster.
The `KafkaNodePool` itself is never deployed outside the central cluster.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  annotations:

>     strimzi.io/stretch-cluster-id: "cluster3"

Reviewer 4: Our requirement here is to uniquely identify a Kubernetes cluster within the [ClusterSet](https://multicluster.sigs.k8s.io/api-types/cluster-set/), but since this is a user-defined and therefore likely human-readable value then `alias` seems appropriate.

In addition, the cluster operator normally creates network policies to restrict traffic from outside the namespace.
To enable pod-to-pod communication across Kubernetes clusters, the cluster operator will not create any network policies when configured to deploy a stretch cluster.

Furthermore, depending on the listener configuration, the Kafka CR may also create LoadBalancer, Ingress, NodePort, and Route services to expose Kafka externally.

Operators deployed to remote clusters are only responsible for reconciling `StrimziPodSet` resources that are created remotely by the operator running in the central cluster.

This approach will allow a Kafka administrator to manage the definition of their stretch cluster in a single location (control plane).

##### Configuration of Kafka properties in a stretch cluster

In a non-stretch (single Kubernetes cluster) Kafka deployment, the `advertised.listener` of a broker pod typically resolves to the pod's internal DNS address:

```yaml

> <broker-pod-name>.<broker-service-name>.<namespace>.svc.cluster.local

Author: You're right that Strimzi typically uses the standard Kubernetes DNS suffix (`.svc`) for intra-cluster communication. However, in the context of stretch cluster deployments, we introduce the `.svc.clusterset.local` suffix deliberately to enable cross-cluster DNS resolution via Kubernetes Multi-Cluster Services (MCS).

The `.clusterset.local` domain is defined by the K8s MCS specification to provide DNS-based service discovery across clusters, and it's required when we want a pod in one cluster to resolve services in another cluster via standard Kubernetes DNS. Without using `.clusterset.local`, services like `<pod-name>.<service>.<namespace>.svc` will only resolve locally within the cluster and will not be routable from remote clusters.

> <broker-pod-name>.<broker-service-name>.<namespace>.svc.cluster.local

Author: Using `.svc` would not allow brokers or controllers in remote clusters to resolve each other. Using `.svc.clusterset.local` ensures cross-cluster service discovery works correctly using DNS. This change is scoped specifically to stretch cluster mode, and single cluster deployments will continue to use `.svc` as before.

> <broker-pod-name>.<broker-service-name>.<namespace>.svc.cluster.local

Author: committed https://github.com/strimzi/proposals/pull/129/commits/d42d8a1f25a3970433c6a5f5eafb7ba872353551

> <broker-pod-name>.<broker-service-name>.<namespace>.svc.cluster.local

Reviewer 2: Please keep in mind that this comment was not about `.clusterset.local`. It was about `.cluster.local` claim made for non-stretched clusters which is not true.

BTW: Where in the MCP spec is the hardcoded definition of `.clusterset.local`? The reason we do not use `cluster.local` is that while this might be the defualt in most clusters, it is configurable and cannot be relied upon.

> <broker-pod-name>.<broker-service-name>.<namespace>.svc.cluster.local

Author: Sorry I'm a bit confused here!

I have created a normal Kafka cluster and this is what I can see

```yaml
controller.quorum.voters=3@my-cluster-controller-3.my-cluster-kafka-brokers.strimzi-test.svc.cluster.local:9090,4@my-cluster-controller-4.my-cluster-kafka-brokers.strimzi-test.svc.cluster.local:9090,5@my-cluster-controller-5.my-cluster-kafka-brokers.strimzi-test.svc.cluster.local:9090
```

The advertised.listeners looks like 

```yaml
advertised.listeners=REPLICATION-9091://my-cluster-broker-0.my-cluster-kafka-brokers.strimzi-test.svc:9091,PLAIN-9092://my-cluster-broker-0.my-cluster-kafka-brokers.strimzi-test.svc:9092,TLS-9093://my-cluster-broker-0.my-cluster-kafka-brokers.strimzi-test.svc:9093
```

I think we have updated our proposal to say the same: https://github.com/strimzi/proposals/pull/129/files#diff-84aeb0d84ac069bfbb02a8de22fe6aa1f24bbc00ac4c9587813726e2e1886e2fR282

> <broker-pod-name>.<broker-service-name>.<namespace>.svc.cluster.local

Author: Yes, `.clusterset.local` is indeed the standardized DNS zone for MCS and is **not configurable as per the current MCS API specification.**

The [KEP-1645](https://github.com/kubernetes/enhancements/blob/master/keps/sig-multicluster/1645-multi-cluster-services-api/specification.md#24---records-for-a-multicluster-headless-service) (section 2.4) explicitly states:

_`“The ClusterSet DNS suffix MUST be clusterset.local.”
"...as this may become configurable in the future, this specification refers to it by the placeholder <clusterset-zone>, but per the MCS API it currently must be defined to be clusterset.local."`_

This means `.clusterset.local` is currently hardcoded and mandated by the API, unlike `.cluster.local`, which is configurable in CoreDNS. So from an implementation standpoint, there is no ambiguity or need to account for variability at this time.

That said, we’re happy to adapt in the future if the standard evolves, for example, by making the zone configurable in the operator, but for now, using `.clusterset.local` is both safe and in line with the spec.

> <broker-pod-name>.<broker-service-name>.<namespace>.svc.cluster.local

Reviewer 2: > Sorry I'm a bit confused here!
> 
> I have created a normal Kafka cluster and this is what I can see
> 
> ```yaml
> controller.quorum.voters=3@my-cluster-controller-3.my-cluster-kafka-brokers.strimzi-test.svc.cluster.local:9090,4@my-cluster-controller-4.my-cluster-kafka-brokers.strimzi-test.svc.cluster.local:9090,5@my-cluster-controller-5.my-cluster-kafka-brokers.strimzi-test.svc.cluster.local:9090
> ```
> 
> The advertised.listeners looks like
> 
> ```yaml
> advertised.listeners=REPLICATION-9091://my-cluster-broker-0.my-cluster-kafka-brokers.strimzi-test.svc:9091,PLAIN-9092://my-cluster-broker-0.my-cluster-kafka-brokers.strimzi-test.svc:9092,TLS-9093://my-cluster-broker-0.my-cluster-kafka-brokers.strimzi-test.svc:9093
> ```
> 
> I think we have updated our proposal to say the same: https://github.com/strimzi/proposals/pull/129/files#diff-84aeb0d84ac069bfbb02a8de22fe6aa1f24bbc00ac4c9587813726e2e1886e2fR282

So as you can see, the advertised listeners do not have the DNS suffix which is in contrast to the proposal.

> <broker-pod-name>.<broker-service-name>.<namespace>.svc.cluster.local

Author: Ahh I see what you mean, updated the proposal to fix that, thanks.

```

This DNS name resolves to the pod IP address within the same Kubernetes cluster. 
Other pods in the cluster can easily reach this pod IP using Kubernetes' internal DNS and networking.

However, in a stretched Kafka cluster, brokers are distributed across multiple Kubernetes clusters.
In this case, a broker's `advertised.listener` must not only identify the broker pod but also indicate which Kubernetes cluster the pod is running in, so that cross-cluster pod-to-pod communication can happen correctly.
To enable this, the operator modifies the broker's advertised listener by inserting the stretch cluster identifier (`<stretch-cluster-id>`) into the DNS name:

```yaml

> <broker-pod-name>.<stretch-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local

Author: The operator determines whether a Kafka deployment is running in stretch mode by checking for the presence of the `STRIMZI_REMOTE_KUBE_CONFIG` env var. If this variable is set, the operator adjusts certain broker configurations such as `advertised.listeners` and `controller.quorum.voters` to use DNS addresses with the `.clusterset.local` suffix. This logic is shown in [this sample POC implementation](https://github.com/aswinayyolath/strimzi-kafka-operator/commit/b79389db605f615ade42495970f747384675265b#diff-ec200265bddfefcd09b5e0be1f9818dffa89fbb0f296c9267c089c05bc0ac63bR227-R237).

The `.clusterset.local` suffix is defined by the K8s MCS API. It enables cross-cluster DNS-based service discovery, which is required in stretch deployments where brokers and controllers are distributed across clusters. Technologies like Submariner and Cilium Cluster Mesh rely on this suffix to resolve services in other clusters within the same multicluster environment.

The full DNS name follows Kubernetes' DNS resolution rules for headless services and StatefulSets. For example:

```yaml
<pod-name>.<stretch-cluster-id>.<service-name>.<namespace>.svc.clusterset.local
```

This name is not directly assigned to the Pod. Instead each broker is part of a headless broker service. The operator knows the pod name, svc name, namespace, and the stretch cluster ID. It constructs the FQDN manually, as shown in PoC implementation. The MCS implementation ensures that this constructed name is resolvable from any cluster in the mesh.

This approach ensures stable and predictable cross-cluster connectivity in stretch mode, without requiring changes to Kubernetes' native DNS mechanisms.

> <broker-pod-name>.<stretch-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local

Author: I think I have answered `Where in the MCS spec is the hardcoded definition of .clusterset.local??` [here](https://github.com/strimzi/proposals/pull/129#discussion_r2074732938)

The `STRIMZI_REMOTE_KUBE_CONFIG` env var is intended to be set only **in the operator running in the central cluster**. This operator is responsible for determining whether the Kafka deployment is in stretch mode and, if so, for creating all necessary resources (such as `StrimziPodSet`) in the remote clusters.

Operators running in the remote clusters do not need to check for `STRIMZI_REMOTE_KUBE_CONFIG`. Instead, the central operator annotates the remote `SPS` with `strimzi.io/remote-podset: true` while creating it. This annotation signals to the operator in the remote cluster that it should reconcile the podset even in the absence of the `Kafka` and `KafkaNodePool` CRs in that cluster. Everything including Secrets, ConfigMaps, Services, etc. is created in the remote cluster by the central CO. The operator in the central cluster will also make all the necessary modifications to `controller.quorum.voters`, `advertised.listeners`, inject additional SANs into broker and controller certs, etc.

Regarding your comment "_This surely needs to be done on a per-cluster basis and not per-operator basis_" ---> I interpret that to mean there is an edge case — when the operator is installed in all-namespaces mode (e.g., in `openshift-operators`) and is used to manage Kafka across multiple namespaces. In that case, if a user tries to deploy both stretch and non-stretch clusters across different namespaces, the global nature of the env var could become a limitation. However, given the fundamental differences in these two deployment modes, we consider it acceptable that they **should not co-exist within the same operator instance**. Users can always run separate operator instances to support this scenario if needed.

> <broker-pod-name>.<stretch-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local

Reviewer 2: This feels like a hidden trap. You hardcode it now and it stops working in the next release. Not your fault obviously, as it is in the MCS spec. But this is worth mentioning explicitly in the proposal.

> <broker-pod-name>.<stretch-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local

Reviewer 2: > The STRIMZI_REMOTE_KUBE_CONFIG env var is intended to be set only in the operator running in the central cluster. This operator is responsible for determining whether the Kafka deployment is in stretch mode and, if so, for creating all necessary resources (such as StrimziPodSet) in the remote clusters.

So, once again, this will not live in isolation. So similarly to why the remote operator might need to be able to operate other operands and not be required as the PodSets only, the central operator might need to run other operands and the node cannot be defined by this environment variable but on the cluster level. I do not think it is acceptable to always require a different operator instance.

> <broker-pod-name>.<stretch-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local

Author: Hmm, thinking aloud here...

I agree that a global env  var like `STRIMZI_REMOTE_KUBE_CONFIG` alone should not dictate whether the operator treats all Kafka deployments as stretch clusters. That would indeed be  problematic especially in a multi-tenant setup or in an all-namespaces operator deployment where stretch and non-stretch clusters may need to co-exist.

But I’d prefer to avoid modifying the Kafka or KNP spec if possible, since that would require a CRD change.  Keeping the CRD backward compatible is a strong advantage, and for a feature like stretch clusters which will likely be opt-in  having the flexibility to evolve without locking in spec design is ideal.

So IMO this is how it should work

Introduce a new annotation like

```yaml

metadata:
  annotations:
    strimzi.io/stretch-cluster: "true"
```

If and only if

1. This annotation is present on the Kafka CR,
2. And the operator has the STRIMZI_REMOTE_KUBE_CONFIG env var set,

...then the operator enables the stretch deployment behavior for that particular Kafka CR:

- It adjusts advertised.listeners, controller.quorum.voters, and FQDNs accordingly,
- It creates SPS resources in remote clusters,
- It injects SANs in cert as required,
-  owns the creation of dependent resources like svc, Secrets, etc. in the remote clusters.
- ...........
- ...........

If the annotation is **not present**, the operator treats the Kafka CR as a normal single-cluster deployment — regardless of the presence of STRIMZI_REMOTE_KUBE_CONFIG.

Why annotation-based control is preferable is because

- no CRD change required.
- supports mixing stretch and non-stretch clusters within a single operator instance.
- operator can emit clear errors if the annotation is used without a valid `STRIMZI_REMOTE_KUBE_CONFIG`.

> <broker-pod-name>.<stretch-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local

Reviewer 2: Using some enable annotation sounds reasonabler. And we already have it for KRaft and Node Pools. So I think this should work.

> <broker-pod-name>.<stretch-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local

Author: Thank you, updated the proposal to address this in [this](https://github.com/strimzi/proposals/pull/129/commits/c1789da642b8a26906ff0ee006ed75d65329431d) commit

```

Here `<stretch-cluster-id>` is a unique identifier for each participating Kubernetes cluster.
This allows the DNS name to be resolvable across clusters using Multicluster networking technologies.
The corresponding IP address resolves to the correct pod IP even if it resides in a different Kubernetes cluster.
The `<stretch-cluster-id>` is dynamically retrieved by the operator from an [annotation](#kafkanodepool-cr) on the `KafkaNodePool` resource.

> The `<stretch-cluster-id>` corresponds to the identifier defined by the user when [configuring Multicluster Services](#step-1-user-configuration-of-the-chosen-mcs-api-implementation), ensuring that brokers are correctly registered in the stretch cluster without requiring manual configuration.

Author: @mark-VIII has pushed a commit https://github.com/strimzi/proposals/pull/129/commits/585d6d28ebf4f9bd627e6c24ec323665c46fed84 to explain in a bit more depth about MCS.

> The `<stretch-cluster-id>` corresponds to the identifier defined by the user when [configuring Multicluster Services](#step-1-user-configuration-of-the-chosen-mcs-api-implementation), ensuring that brokers are correctly registered in the stretch cluster without requiring manual configuration.

Author: I'm marking this thread as resolved we can open this if you think this need bit more explanation.

PLAIN-9092://<broker-pod-name>.<stretch-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local:9092,
TLS-9093://<broker-pod-name>.<stretch-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local:9093
```

Similarly, the `controller.quorum.voters` configuration is updated to reference controllers across all participating clusters:

```yaml
controller.quorum.voters=<controller-id>@<controller-pod-name>.<stretch-cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local:9090, ...
```

These updates ensure brokers and controllers can be discovered and communicate across Kubernetes clusters without relying on traditional external access methods.

#### Prototype

A working prototype can be deployed using the steps outlined in a draft [README](https://aswinayyolath.github.io/stretch-kafka-docs/) that is being iteratively revised.

_Note: The prototype might not always exactly align with this proposal so please refer to the `README` documentation when working with the prototype._

### Additional considerations and reference information

> #### Authentication and security considerations for multi-cluster access

Author: In stretch cluster mode, Kafka communication remains secured via TLS, and hostname verification using `ssl.endpoint.identification.algorithm=HTTPS` continues to function without changes. To support this, the operator adds additional SAN entries in the broker and controller certificates for the `.clusterset.local` DNS names used in cross-cluster commuication.

For eg, in stretch mode there'll be SANs like

```yaml
DNS:my-cluster-broker-0.cluster1.my-cluster-kafka-brokers.stretch.svc.clusterset.local
DNS:my-cluster-controller-5.cluster1.my-cluster-kafka-brokers.stretch.svc.clusterset.local
```

These entries ensure that the TLS handshake passes hostname verification when traffic flows across clusters. This is the only additional change related to authentication in stretch mode - no modifications to the TLS mechanisms or Kafka's existing security model are required.

PoC implementation for SAN injection can be seen here: https://github.com/aswinayyolath/strimzi-kafka-operator/commit/c8c4cbef10f4ef80240388a575431a00d39a8139

> #### Authentication and security considerations for multi-cluster access

Reviewer 2: Right, so add it to the proposal.

> #### Authentication and security considerations for multi-cluster access

Author: I have already added it yesterday but forgot to mention in the comment !

###### Recovery approaches
There are several possible recovery methods:

**1. Manual recovery (baseline approach)**
Users bring up a new Kubernetes cluster and manually apply the `Kafka` and `KafkaNodePool` CRs.
Alternatively, an existing remote cluster can be promoted to central by applying these CRs and updating the cluster operator deployment.
This approach is simple and reliable but requires user intervention.

**2. GitOps based recovery**

Users store CRs in GitOps tools (e.g., ArgoCD, Flux).
When a new central cluster is created, GitOps tools automatically reconcile the resources.
This reduces manual effort but requires GitOps workflows to be set up in advance.

#### Feature Enablement and Rollout Plan
The Stretch Kafka Cluster support will be introduced as an optional feature behind a feature gate called `UseStretchCluster`.
The maturity and rollout of this feature gate will follow the standard Strimzi process:
- The feature gate will be disabled by default, allowing early adopters and community members to safely test the functionality without affecting production environments.

> - After at least two Strimzi releases, and based on user feedback and observed stability, enabling the feature gate by default may be considered.

Reviewer 5: I suggested the use of a feature gate. Since this is a large and likely complicated feature to implement, I think we should use the feature gate to let the feature "bake" for a release or so before marking it production ready.

> - After at least two Strimzi releases, and based on user feedback and observed stability, enabling the feature gate by default may be considered.

Reviewer 2: So, if we have a feature gate, it should have a clear timeline. I think this should be pretty hard to enable and should have a lot of opportunities to tell the user at what stage of development it is even without the feature gate (the feature gate here will serve only as an acknowledgement that users are aware of the state, nothing technical). But I guess if you want to have a feature gate it is fine with me.

> - After at least two Strimzi releases, and based on user feedback and observed stability, enabling the feature gate by default may be considered.

Reviewer 3: I agree on having this behind a feature gate despite to enable a stretch cluster needs some steps and configuration. A FG makes clear to the users that's a beta feature to be tested before maturing. Of course, we would need a timeline, I agree.

Existing Kafka cluster deployments will remain unaffected unless users explicitly enable the UseStretchCluster feature gate and configure the necessary settings.

## Affected/not affected projects

This proposal only impacts the strimzi-kafka-operator project.

### Entity operator
We recommend that all `KafkaTopic` and `KafkaUser` resources are managed from the central cluster: this is also where the entity operator should be deployed.
This will maintain the central cluster as the single control plane for resource management.
The expectation is that the entity operator will not be impacted any further by changes made to support a stretch cluster.

### Kafka Connect, Kafka Bridge and MirrorMaker2
This proposal does not cover stretching Kafka Connect, Kafka MirrorMaker 2 or the Kafka Bridge.
These components will be deployed to the central cluster and will function as they do today.

> Operators running in remote clusters will not manage KafkaBridge, KafkaConnect, KafkaConnector, or KafkaMirrorMaker2 resources.

Reviewer 5: I think we at least need a little more explanation here. For example, presumably a KafkaConnect cluster must now be given an external address as the bootstrap address? Also stating `operators running in remote clusters will not manage KafkaBridge..` is not precisely true. Regardless of whether the operator managing the stretch cluster is reconciling these other resources, the user could deploy a separate operator in the remote cluster to handle them.

So I think as I stated in the thread about the `STRIMZI_POD_SET_RECONCILIATION_ONLY` env var, really we do need to be clear about how we expect users to make use of a stretch cluster alongside these other components.

> Operators running in remote clusters will not manage KafkaBridge, KafkaConnect, KafkaConnector, or KafkaMirrorMaker2 resources.

Reviewer 3: As in the same thread I left a comment about the possibility to avoid the `STRIMZI_POD_SET_RECONCILIATION_ONLY` and having remote cluster operator operating all other components deployed locally. For the stretch cluster it would make sense to continue to deploy in the central one imho.

### Limitations and Considerations
While a stretch Kafka cluster offers several advantages, it also introduces some challenges and considerations:

- **Increased Network Complexity and Costs**: The communication between brokers and controllers across clusters relies on network connectivity, which can be less reliable and more costly than intra-cluster communication.
This necessitates careful consideration of network architecture and associated costs.

- **Latency Requirements**: Stretch Kafka clusters are best suited for environments with low-latency network connections between the Kubernetes clusters.
High latency can adversely affect the performance and synchronization of Kafka nodes, potentially leading to delays or errors in replication and client communication.
Defining the minimal acceptable latency between clusters is crucial to ensure optimal performance.

## Proposal

This proposal seeks to enhance the Strimzi Kafka operator to support stretch Kafka clusters, distributing brokers and controllers across multiple Kubernetes clusters.

> The intent is to focus on high-availability of the data plane.

Reviewer 6: Yes, in the context of the proposal, the control plane meant the central cluster and its operator, which manages the rest of the clusters, and the data plane meant the clusters where the central operator deploys the Kafka pods (the clusters other than the central one).

I agree that this is causing confusion, as in the Strimzi context, the control plane constitutes the controller pods, and the data plane constitutes the broker pods. We do mean that the controllers and brokers should be deployed in all the clusters for HA. We've updated the proposal to have clarified the definition of control plane added to this.

What the proposal is trying to articulate is that even if the central cluster, which contains all the CRs, goes down, clients will still be served by the Kafka brokers and controllers running in the remaining clusters.

The proposal outlines high-level topology and design concepts for such deployments, with a plan to incrementally include finer design and implementation details for various aspects.

### Prerequisites

- **Multiple Kubernetes clusters**: Stretch Kafka clusters will require multiple Kubernetes clusters.
To ensure Kafka controller quorum in the event of a Kubernetes cluster outage, it is recommended to distribute controllers across at least three Kubernetes clusters. In a two cluster setup, there is a risk of losing quorum if the wrong cluster fails, which can lead to Kafka becoming unavailable.

- **Low Latency**: Kafka clusters should be deployed in environments that allow low-latency communication between Kafka brokers and controllers.
Stretch Kafka clusters should be deployed in environments such as data centers or availability zones within a single region, and not across distant regions where high latency could impair performance.

> - **KRaft**: As Kafka and Strimzi transition towards KRaft-based clusters, this proposal focuses exclusively on enabling stretch deployments for KRaft-based Kafka clusters.

Reviewer 6: Agreed, I've moved this point to the top.

    The MCS API is not used by the prototype, but manual re-configuration of Kubernetes CoreDNS is required.

    The draft [README](https://aswinayyolath.github.io/stretch-kafka-docs) for the prototype includes detailed steps describing how to pre-configure both [Submariner](https://aswinayyolath.github.io/stretch-kafka-docs/setting-up-submariner/) and [Cilium](https://aswinayyolath.github.io/stretch-kafka-docs/Setting-up-cilium/).

### Design

#### Topology of a stretch cluster

![Stretch cluster topology](./images/095-stretch-cluster-topology.png)

The diagram illustrates a topology comprising of three Kubernetes clusters.

One of these clusters is designated as the "Central cluster", while any additional clusters are considered "remote".
The central cluster acts as the control plane where a user will create all the custom resources for the Kafka cluster - Kafka, KafkaNodePool, KafkaUser, KafkaTopic etc.

A Kafka node pool definition can be configured to specify a Kubernetes cluster (central cluster or one of the member clusters) as the deployment target.

> The operator on the central cluster is responsible for creating all necessary resources (including `StrimziPodSet` resources) for the node pool on the target Kubernetes cluster. 

Author: Even if the central cluster goes down, clients will continue to be served since brokers and controllers are distributed across multiple Kubernetes clusters. To recover, users would need to bring up a new Kubernetes cluster and apply the Kafka ,KafkaNodePool CRs and CO env vars there. Alternatively, they could apply these CRs and update the CO deployment in an existing K8s cluster to designate it as the new central cluster.

The current assumption is that a central cluster outage does not impact Kafka clients, as the brokers remain operational. We have conducted preliminary tests to validate this, but we will perform further testing to confirm and update the findings here.

Regarding the idea of replicating cluster related resources (including KafkaNodePool) across remote clusters and having them "silently" ignored unless the operator is running in central  is certainly a viable approach, but it introduces additional complexity. We would appreciate expert insights from experienced maintainers like you on the feasibility and trade-offs of this approach.

> The operator on the central cluster is responsible for creating all necessary resources (including `StrimziPodSet` resources) for the node pool on the target Kubernetes cluster. 

Reviewer 6: Our PoC ensures resilience against a central cluster failure. If the central cluster goes down, the remaining clusters continue operating as expected. Any partition leaders residing in the failed central cluster are automatically re-elected to brokers in the running clusters. Detailed test results can be found [here](https://aswinayyolath.github.io/stretch-kafka-docs/Testing-cluster-failover/).
To maximize recoverability, rack awareness can be configured to ensure that at least one replica is present in each cluster. This guarantees that as long as at least one Kubernetes cluster remains operational, no data is lost.

> The operator on the central cluster is responsible for creating all necessary resources (including `StrimziPodSet` resources) for the node pool on the target Kubernetes cluster. 

Reviewer 7: Thanks @aswinayyolath and @rohan-anilkumar, it makes sense. It would be good to capture this in a disaster recovery section of the proposal.

> The operator on the central cluster is responsible for creating all necessary resources (including `StrimziPodSet` resources) for the node pool on the target Kubernetes cluster. 

Author: That’s a great suggestion! We’ll make sure to capture this in the DR section of the proposal to clearly outline the approach and its impact on resilience. Thanks for the feedback!

> The operator on the central cluster is responsible for creating all necessary resources (including `StrimziPodSet` resources) for the node pool on the target Kubernetes cluster. 

Author: @fvaleri @ppatierno I have added [this](https://github.com/strimzi/proposals/pull/129/commits/d8e2f3162a2213842dbf48650f69f1b2285d7434) commit to address this. appreciate your review.

> The operator on the central cluster is responsible for creating all necessary resources (including `StrimziPodSet` resources) for the node pool on the target Kubernetes cluster. 

Reviewer 3: I wasn't referring to the behaviour respect to the Kafka clients. I assume Kafka is going to do its job :-)
I was referring to the disaster recovery of the central operator. If it fails, there is no sort of mechanism of an election having one of the remote clusters to become the central one, but instead you suggested the user to create a new Kube cluster and copy over the resources there and configuring it as central.
It means that Kafka clients will use the cluster with no issues, but the Kafka cluster itself remain un-operated for some time because no operator is looking at changes in the Kafka or KafkaNodePool resources (i.e. storage change, config change, ....).

> The operator on the central cluster is responsible for creating all necessary resources (including `StrimziPodSet` resources) for the node pool on the target Kubernetes cluster. 

Author: @ppatierno 

You're absolutely right, while Kafka clients remain unaffected, the Kafka cluster itself would be temporarily unoperated in the event of a central cluster failure.

Currently, there is no automated failover mechanism for the central operator. If the central cluster goes down, users would need to either

1. Bring up a new Kubernetes cluster, apply the Kafka and KafkaNodePool CRs, and configure the CO deployment as central.
2. Promote an existing cluster by applying the necessary CRs and updating the CO deployment to take over the central role.

Introducing a mechanism where remote clusters can detect a central cluster failure and elect one among them to take over as central would add resilience but also significant complexity, especially in maintaining state consistency across clusters.

One possible approach, as you suggested, is to replicate Kafka cluster-related resources (KafkaNodePool, etc.) across remote clusters so they are readily available if a failover is needed. However, this requires careful handling to prevent unintended actions when the original central cluster is operational.

We would love to hear your thoughts on how best to balance resilience and complexity here. Would you see value in an automated failover approach, or do you think a more manual recovery process is reasonable given the distributed nature of the setup?

> The operator on the central cluster is responsible for creating all necessary resources (including `StrimziPodSet` resources) for the node pool on the target Kubernetes cluster. 

Reviewer 3: I see your concerns about replicating Kafka and KafkaNodePool resources across the remote clusters which would just speed up the process of restoring a central cluster but it's more keen to make mistakes.
Let's take apart any automation for now but I think this manual approach should be described in the proposal.

> Operators deployed to remote clusters are only responsible for reconciling `StrimziPodSet` resources that are created remotely by the operator running in the central cluster.

Author: That's a great point, and it's definitely worth investigating. Currently, the remote operator is only responsible for reconciling SPS resources, so limiting it to run only the StrimziPodSetController could indeed be a more resource-efficient approach. However, we would need to evaluate whether Strimzi's architecture allows for selectively enabling or disabling specific controllers at runtime without unintended side effects. If feasible, this could help optimize resource usage in remote clusters while ensuring smooth reconciliation. We can explore this further and assess the potential impact.

> Operators deployed to remote clusters are only responsible for reconciling `StrimziPodSet` resources that are created remotely by the operator running in the central cluster.

Author: @ppatierno we investigated this and tested running the operator in a "stretch mode" by setting `STRIMZI_POD_SET_RECONCILIATION_ONLY=true` in the remote cluster operator deployment. This ensured that only the StrimziPodSetController started, while other controllers (Kafka, KafkaConnect, etc.) remained inactive.

We verified that Kafka pods were successfully created and running across all Kubernetes clusters, and everything functioned as expected. We've also [documented](https://aswinayyolath.github.io/stretch-kafka-docs/Deploying-Strimzi-in-Stretch-Mode/#updating-environment-variables) this approach.

> Operators deployed to remote clusters are only responsible for reconciling `StrimziPodSet` resources that are created remotely by the operator running in the central cluster.

Reviewer 3: Cool! But how does it cope with the `STRIMZI_STRETCH_MODE` as true? I mean, one implies the other at this point, or?

> Operators deployed to remote clusters are only responsible for reconciling `StrimziPodSet` resources that are created remotely by the operator running in the central cluster.

Author: updated proposal addressed this

_Note: This network identifier could have the same value as the Kubernetes cluster identifier described in the following section and both aspects would benefit from community feedback and input._

#### User configuration of the deployment topology

The operator running in the central Kubernetes cluster must be provided with the following information to allow creation of resources in remote clusters:

- An identifier for each remote Kubernetes cluster.
- A URL endpoint for the Kubernetes API server running in each remote cluster.
- Credential(s) to allow authentication with the remote Kubernetes API servers.

For simplicity and to quickly achieve a working prototype, the information outlined above is currently provided as an environment variable to the central operator.
The value of the environment variable uses a "map" format as shown below:

```yaml
- name: STRIMZI_K8S_CLUSTERS
  value: |
      cluster-id-a.url=<cluster-a URL>
      cluster-id-a.secret=<secret-name-cluster-a>
      cluster-id-b.url=<cluster-b URL>

>       cluster-id-b.secret=<secret-name-cluster-b>

Reviewer 6: The user should create secrets containing the kubeconfig credentials in the central cluster. The secret name and cluster endpoint URL should then be added as environment variables in the cluster operator deployment. The operator code will then parse this secret to establish connections to the remote clusters and create/manage resources in them.

The code for this can be found in our POC [here](https://github.com/aswinayyolath/strimzi-kafka-operator/compare/main...aswinayyolath:strimzi-kafka-operator:strecth-cluster-prototype?expand=1#diff-ac9b17492d6e343d1a402e70ab4460b2f56e6d3a66600f707cb354523384c651R255)

>       cluster-id-b.secret=<secret-name-cluster-b>

Reviewer 3: Can we clarify, within the proposal, how these Secrets are used?

>       cluster-id-b.secret=<secret-name-cluster-b>

Author: Yeah sure, I will update the proposal to clarify this

The prototype uses a `spec.cluster` field within a KafkaNodePool definition that will have a value that matches one of the identifiers from the above map.
If the `spec.cluster` field is missing from the KafkaNodePool definition, it is currently assumed that the broker/controller pods will run on the central Kubernetes cluster.

_Note: This cluster identifier could have the same value as the network identifier described in the previous section. Community feedback/input very welcome!_

#### Other prototype configuration items and limitations

1. The central cluster operator needs to know which of the Cloud Native Network projects has been pre-configured. This is currently set on the Kafka CR using an annotation:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: Kafka
metadata:
  ...
  annotations:
    ...
    strimzi.io/cross-cluster-type: "submariner"
```

2. All Strimzi cluster operators involved must be aware they are involved in reconciling a stretch Kafka cluster by means of a new environment variable:

```yaml
- name: STRIMZI_STRETCH_MODE
  value: 'true'
```

> Within the prototype operator, this is used to conditionally break the expectation that Kafka, KafkaNodePool are located in the same Kubernetes cluster as the StrimziPodSet and Pod resources.

Author: Currently, `STRIMZI_STRETCH_MODE` is used to break the assumption that Kafka and KafkaNodePool must be present in the same cluster where StrimziPodSet is reconciled. In the existing Strimzi implementation, both Kafka and KafkaNodePool are required to bring up Kafka Pods. However, in a stretch cluster setup, these CRs are only present in the central cluster, while Pods need to be created in remote clusters without relying on the presence of Kafka or KafkaNodePool CRs there. The `STRIMZI_STRETCH_MODE` environment variable enables this behavior by allowing the operator to handle reconciliation differently for stretch clusters.

Currently, the distinction between the central cluster and remote clusters is determined in two ways:

- Checking the Cluster Operator deployment —> if it has the `STRIMZI_K8S_CLUSTERS` environment variable set, it is running in the central cluster.
- Checking whether a Kafka or KafkaNodePool CR exists in the cluster —> these CRs are only applied in the central cluster.

The operator deploys pod in central or remote cluster by looking at the `cluster` property in the KafkaNodePool CR(an example is given [here](https://aswinayyolath.github.io/stretch-kafka-docs/Deploying-Strimzi-in-Stretch-Mode/#applying-kafka-and-kafkanodepool-crs-in-the-central-cluster)). If this property is present, it indicates a remote cluster; if absent, the CR belongs to the central cluster. To improve clarity, we could consider renaming this property (e.g., to remote-cluster) or moving it to annotations/labels instead of keeping it inside the KafkaNodePool spec.

Additionally, if STRIMZI_STRETCH_MODE is not set, Strimzi operates as it does today, following the standard single Kubernetes cluster deployment model without any changes to its existing behavior.

I hope this explanation makes sense. Let me know your thoughts!

> Within the prototype operator, this is used to conditionally break the expectation that Kafka, KafkaNodePool are located in the same Kubernetes cluster as the StrimziPodSet and Pod resources.

Reviewer 3: @aswinayyolath thanks for the explanation but it was already clear to me :-) My feeling is that the `STRIMZI_STRETCH_MODE` env var name is kind of misleading. If it's set to "true" if you kind of referring to the "remote" Strimzi operator to deal with just StrimziPodSets creation (because Kafka and KafkaNodePool are in the central one). But If I put myself from the central operator perspective, even this one is working in stretch mode so `STRIMZI_STRETCH_MODE` should be true as well, because it's handling pools with pods running elsewhere.
I hope my reasoning is clear now. This is why I was suggesting to replace it with an env var defining the role/mode of the operator and it's set on both central and remote.

> Within the prototype operator, this is used to conditionally break the expectation that Kafka, KafkaNodePool are located in the same Kubernetes cluster as the StrimziPodSet and Pod resources.

Author: Thanks for the feedback! I completely get your reasoning, and that makes sense. In the revised proposal, `STRIMZI_STRETCH_MODE` is no longer there.

When Kafka is deployed using "stretch mode", the operator modifies `advertised.listeners` to include a network identifier, such as the Submariner or Cilium `cluster-id`. The modified format is:

```
advertised.listeners=REPLICATION-9091://<broker-pod-name>.<cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local:9091,
PLAIN-9092://<broker-pod-name>.<cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local:9092,
TLS-9093://<broker-pod-name>.<cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local:9093
```

The `<cluster-id>` is dynamically retrieved from a label on the `KafkaNodePool` resource.

**Modified `controller.quorum.voters` format**

Similarly, `controller.quorum.voters` includes all Kafka controllers running in all of the Kubernetes clusters involved in the stretch Kafka cluster:

```
controller.quorum.voters=<controller-id>@<controller-pod-name>.<cluster-id>.<broker-service-name>.<namespace>.svc.clusterset.local:9090, ...
```

These updates ensure brokers and controllers can be discovered and communicate across clusters without relying on traditional external access methods.

#### Resource cleanup on remote Kubernetes clusters

> Currently, resources created in the remote clusters do not have `OwnerReferences`. The main reason is that the `Kafka` and `KafkaNodePool` CRs exist only in the central cluster. If resources in remote clusters were to reference them as owners, they would be immediately deleted by Kubernetes' garbage collection mechanism since their owners do not exist in the remote clusters.

Reviewer 6: Yes. The POC is what we meant in this context, I've updated this section to give more clarity to this.

# Stretch Kafka cluster

The Strimzi Kafka operator currently manages Kafka clusters within a single Kubernetes cluster.
This proposal aims to extend this support to stretch Kafka clusters, where brokers and controllers of a single Kafka cluster are distributed across multiple Kubernetes clusters.

## Current situation

At present, the availability of Strimzi-managed Kafka clusters is limited by the availability of the underlying Kubernetes cluster.

> If a Kubernetes cluster experiences an outage, the entire Kafka cluster becomes unavailable, disrupting all connected Kafka clients.

Author: Sure that make sense. will add that

> If a Kubernetes cluster experiences an outage, the entire Kafka cluster becomes unavailable, disrupting all connected Kafka clients.

Author: updated this section to address this

- **Resource Optimization**: Efficiently utilizing resources across multiple clusters, which can be advantageous in environments with varying cluster capacities or during scaling operations.

### Limitations and Considerations
While a stretch Kafka cluster offers several advantages, it also introduces some challenges and considerations:

- **Increased Network Complexity and Costs**: The communication between brokers and controllers across clusters relies on network connectivity, which can be less reliable and more costly than intra-cluster communication.
This necessitates careful consideration of network architecture and associated costs.

- **Latency Requirements**: Stretch Kafka clusters are best suited for environments with low-latency network connections between the Kubernetes clusters.
High latency can adversely affect the performance and synchronization of Kafka nodes, potentially leading to delays or errors in replication and client communication.
Defining the minimal acceptable latency between clusters is crucial to ensure optimal performance.

## Proposal

> This proposal seeks to enhance the Strimzi Kafka operator to support stretch Kafka clusters, distributing brokers and controllers across multiple Kubernetes clusters.

Reviewer 7: ```suggestion
This proposal seeks to enhance the Strimzi Kafka operator to support stretch Kafka clusters, distributing broker, controller and combined Kafka nodes across multiple Kubernetes clusters.
```

> The intent is to focus on high-availability of the control plane and data plane.

Reviewer 7: ```suggestion
The intent is to focus on high-availability of the Kafka control plane and data plane.
```

The proposal outlines high-level topology and design concepts for such deployments, with a plan to incrementally include finer design and implementation details for various aspects.

### Prerequisites

- **KRaft**: As Kafka and Strimzi transition towards KRaft-based clusters, this proposal focuses exclusively on enabling stretch deployments for KRaft-based Kafka clusters.
While Zookeeper-based deployments are still supported, they are outside the scope of this proposal.

- **Multiple Kubernetes clusters**: Stretch Kafka clusters will require multiple Kubernetes clusters.
To ensure Kafka controller quorum in the event of a Kubernetes cluster outage, it is recommended to distribute controllers across at least three Kubernetes clusters. In a two cluster setup, there is a risk of losing quorum if the wrong cluster fails, which can lead to Kafka becoming unavailable.

> - **Low Latency**: Kafka clusters should be deployed in environments that allow low-latency communication between Kafka brokers and controllers.

Reviewer 7: and high-bandwidth?

_Note: This network identifier could have the same value as the Kubernetes cluster identifier described in the following section and both aspects would benefit from community feedback and input._

#### User configuration of the deployment topology

The operator running in the central Kubernetes cluster must be provided with the following information to allow creation of resources in remote clusters:

- An identifier for each remote Kubernetes cluster.
- A URL endpoint for the Kubernetes API server running in each remote cluster.
- Credential(s) to allow authentication with the remote Kubernetes API servers.

For simplicity and to quickly achieve a working prototype, the information outlined above is currently provided as an environment variable to the central operator.
The value of the environment variable uses a "map" format as shown below:

```yaml
- name: STRIMZI_K8S_CLUSTERS
  value: |
      cluster-id-a.url=<cluster-a URL>
      cluster-id-a.secret=<secret-name-cluster-a>
      cluster-id-b.url=<cluster-b URL>
      cluster-id-b.secret=<secret-name-cluster-b>
```

> The secrets referenced here must currently contain the kubeconfig for a the Kubenetes cluster available at the provided URL as the value of secret key 'kubeconfig'.

Reviewer 7: ```suggestion
The secrets referenced here must currently contain the kubeconfig for the Kubenetes cluster available at the provided URL as the value of secret key 'kubeconfig'.
```

#### Entity operator
We recommend that all `KafkaTopic` and `KafkaUser` resources are managed from the central cluster: this is also where the entity operator should be deployed.
This will maintain the central cluster as the single control plane for resource management.
The expectation is that the entity operator will not be impacted any further by changes made to support a stretch cluster.

## Feature Enablement and Rollout Plan
The Stretch Kafka Cluster support will be introduced as an optional feature behind a feature gate called `UseStretchCluster`.
The maturity and rollout of this feature gate will follow the standard Strimzi process:
- The feature gate will be disabled by default, allowing early adopters and community members to safely test the functionality without affecting production environments.
- After at least two Strimzi releases, and based on user feedback and observed stability, enabling the feature gate by default may be considered.

Existing Kafka cluster deployments will remain unaffected unless users explicitly enable the UseStretchCluster feature gate and configure the necessary settings.

> ## Kafka Connect and MirrorMaker2 Considerations

Reviewer 4: @ppatierno I believe you are correct in regard to both Kafka Bridge and Drain Cleaner. We will be adding details to the proposal early next week.

High latency can adversely affect the performance and synchronization of Kafka nodes, potentially leading to delays or errors in replication and client communication.
Defining the minimal acceptable latency between clusters is crucial to ensure optimal performance.

## Proposal

This proposal seeks to enhance the Strimzi Kafka operator to support stretch Kafka clusters, distributing broker, controller and combined Kafka nodes across multiple Kubernetes clusters.
The intent is to focus on high-availability of the Kafka control plane and data plane.
The proposal outlines high-level topology and design concepts for such deployments, with a plan to incrementally include finer design and implementation details for various aspects.

## Prerequisites

- **KRaft**: As Kafka and Strimzi transition towards KRaft-based clusters, this proposal focuses exclusively on enabling stretch deployments for KRaft-based Kafka clusters.
While Zookeeper-based deployments are still supported, they are outside the scope of this proposal.

- **Multiple Kubernetes clusters**: Stretch Kafka clusters will require multiple Kubernetes clusters.

> To ensure Kafka controller quorum in the event of a Kubernetes cluster outage, it is recommended to distribute controllers across at least three Kubernetes clusters. In a two cluster setup, there is a risk of losing quorum if the wrong cluster fails, which can lead to Kafka becoming unavailable.

Author: Strimzi will not enforce a minimum of three clusters but will strongly recommend it. A two-cluster setup introduces a risk of losing quorum if the wrong cluster fails. We could add a warning in the Operator when users configure fewer than three clusters, helping them make an informed decision.

> To ensure Kafka controller quorum in the event of a Kubernetes cluster outage, it is recommended to distribute controllers across at least three Kubernetes clusters. In a two cluster setup, there is a risk of losing quorum if the wrong cluster fails, which can lead to Kafka becoming unavailable.

Reviewer 5: A warning condition similar to the ones when running with a single node makes sense to me

> To ensure Kafka controller quorum in the event of a Kubernetes cluster outage, it is recommended to distribute controllers across at least three Kubernetes clusters. In a two cluster setup, there is a risk of losing quorum if the wrong cluster fails, which can lead to Kafka becoming unavailable.

Author: exactly! That’s what I was thinking. This also allows users to test with a two cluster setup in testing or pre-prod env without needing to provision three clusters just for testing.

> To ensure Kafka controller quorum in the event of a Kubernetes cluster outage, it is recommended to distribute controllers across at least three Kubernetes clusters. In a two cluster setup, there is a risk of losing quorum if the wrong cluster fails, which can lead to Kafka becoming unavailable.

Author: updated the proposal to address this

- **Low Latency and High Bandwidth**:  Kafka clusters should be deployed in environments that provide low-latency and high-bandwidth communication between Kafka brokers and controllers.
Stretch Kafka clusters should be deployed in environments such as data centers or availability zones within a single region.
Deployments across distant regions, where high latency and limited bandwidth could impair performance, should be avoided.

- **A supported cloud native networking technology**: To enable networking between Kubernetes clusters currently requires an additional technology stack.

> The prototype detailed within this proposal requires advance manual setup of an overlay network offered by a [Cloud Native Network](https://landscape.cncf.io/guide#runtime--cloud-native-network) project to provide connectivity between Kafka runtimes in different Kubernetes clusters.

Author: Yes, a Cloud Native networking solution is required for cross cluster communication in stretch Kafka clusters. The final implementation will provide a list of validated networking technologies, starting with Submariner and Cilium. However, users can choose alternative networking solutions as long as they provide the required cross-cluster connectivity. Over time, we expect to validate more options.

> The prototype detailed within this proposal requires advance manual setup of an overlay network offered by a [Cloud Native Network](https://landscape.cncf.io/guide#runtime--cloud-native-network) project to provide connectivity between Kafka runtimes in different Kubernetes clusters.

Author: Strimzi requires direct pod to pod communication for stretch Kafka clusters to function correctly. Brokers must communicate with each other for replication, controllers must coordinate for leader election, and brokers and controllers must synchronize metadata state. Strimzi relies on direct pod to pod communication to facilitate this, rather than routing traffic through services or proxies

To enable this across K8s clusters in a stretch Kafka deployment, we need a K8s native networking solution that supports direct pod to pod connectivity across clusters. In our testing, we successfully achieved this using `Submariner` and `Cilium` Cluster Mesh. These solutions allow Kafka brokers and controllers in different Kubernetes clusters to communicate directly.

We also attempted to use `Istio` and `Skupper`, but we were unable to make them work for this use case. IIRC, these technologies do not provide direct pod to pod connectivity across clusters, or they had limitations that prevented seamless integration with Kafka’s networking requirements.

In theory, any cloud-native networking solution that enables direct pod to pod communication across clusters should work, but so far, only Submariner and Cilium have been successfully tested

> The prototype detailed within this proposal requires advance manual setup of an overlay network offered by a [Cloud Native Network](https://landscape.cncf.io/guide#runtime--cloud-native-network) project to provide connectivity between Kafka runtimes in different Kubernetes clusters.

Author: From Strimzi's perspective, Submariner and Cilium function the same way in a stretch Kafka deployment, even though their setup configurations differ. Both technologies assign a clusterid to each Kubernetes cluster where they are installed. This cluster ID is the primary information required by Strimzi to dynamically generate `controller.quorum.voters` and `advertised.listeners`.

From an Operator implementation standpoint, supporting Submariner and Cilium does not require significant changes. The key requirement is for users to provide the clusteid associated with their chosen networking technology. This can be achieved by using an annotation in the KNP CR, where users specify the clusterid.

there is one notable difference

- Submariner requires an additional CR called ServiceExport, which is used to expose `controller.quorum.voters` and `advertised.listeners` outside of a Kubernetes cluster.
- Cilium does not require this extra step.

To handle this in the Operator, the code should conditionally check which cloud native networking technology is being used. If Submariner is detected, the Operator should automatically create a ServiceExport CR. If Cilium is used, this step is unnecessary.

    The MCS API is not used by the prototype, but manual re-configuration of Kubernetes CoreDNS is required.

    The draft [README](https://aswinayyolath.github.io/stretch-kafka-docs) for the prototype includes detailed steps describing how to pre-configure both [Submariner](https://aswinayyolath.github.io/stretch-kafka-docs/setting-up-submariner/) and [Cilium](https://aswinayyolath.github.io/stretch-kafka-docs/Setting-up-cilium/).

## Design

### Topology of a stretch cluster

![Stretch cluster topology](./images/095-stretch-cluster-topology.png)

The diagram illustrates a topology comprising of three Kubernetes clusters.

One of these clusters is designated as the "Central cluster", while any additional clusters are considered "remote".
The central cluster acts as the control plane where a user will create all the custom resources for the Kafka cluster - Kafka, KafkaNodePool, KafkaUser, KafkaTopic etc.

A Kafka node pool definition can be configured to specify a Kubernetes cluster (central cluster or one of the member clusters) as the deployment target.

> The operator on the central cluster is responsible for creating all necessary resources (including `StrimziPodSet` resources) for the node pool on the target Kubernetes cluster. 

Author: I have added a section detailing the resources that the central cluster needs to create in the remote K8s clusters

This configuration must be performed by the user prior to deployment of Strimzi cluster operators.

Each Kubernetes cluster joined to the network is assigned a unique identifier.
This identifier is used by the prototype central operator to build:
1. Valid broker and controller service endpoints values for `advertised.listeners` and `controller.quorum.voters` within the appropriate ConfigMap resources.
2. Broker certificates with appropriate Subject Alternative Names.
The prototype currently expects this identifier to be supplied by the user as a label value on the Kafka node pool resource:

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  ...
  labels:
    ...
    strimzi.io/submariner-cluster-id: "cluster1"
```

_Note: The prototype operator [code](https://github.com/aswinayyolath/strimzi-kafka-operator/compare/main...aswinayyolath:strimzi-kafka-operator:strecth-cluster-prototype?expand=1) currently expects `submariner-cluster-id`, but this is subject to change for something more generic once additional technologies have been prototyped. One suggestion is `strimzi.io/cluster-network-id`._

> _Note: This network identifier could have the same value as the Kubernetes cluster identifier described in the following section and both aspects would benefit from community feedback and input._

Author: I see how the distinction between the network identifier and the cluster identifier might be unclear, so l'll try to clarify

Network Identifier (`strimzi.io/submariner-cluster-id`) is used to define how Kafka nodes communicate across clusters. It determines values for `controller.quorum.voters` and `advertised.listeners`, ensuring nodes in different Kubernetes clusters can discover and connect to each other. In a Submariner based deployment, this id corresponds to Submariner’s clusterId. If we support other networking technologies like Cilium, we’ll need a similar identifier for them as well. now to generalize this, we’re considering renaming it to somethinh like `strimzi.io/cluster-network-id`

Now the cluster Identifier (`spec.cluster` in KafkaNodePool) tells the operator where to deploy resources (i.e., in which Kubernetes cluster a particular KafkaNodePool should create SPS, Secret,CM,SA,SVC etc). The operator uses this value to find the correct kubeconfig credentials and apply the resources in the corresponding remote cluster. Currently, this is specified in the `spec.cluster` field of KNP, but we could consider using a standard annotation instead (e.g., strimzi.io/cluster-id or something similar) to make configuration more intuitive, or else we can keep only a single lable for both network and cluster identifier why will make it easy for the users as they just need to set one thing in the KNP CR.

In short the network identifier ensures Kafka nodes can communicate across clusters and the cluster identifier ensures the Strimzi operator knows where to create Kafka resources.

> _Note: This network identifier could have the same value as the Kubernetes cluster identifier described in the following section and both aspects would benefit from community feedback and input._

Reviewer 7: Both `cluster-id` and `network-id` seems to be a bit overloaded, should we use something like `strimzi.io/stretch-cluster-id` and `strimzi.io/stretch-network-id`?

> _Note: This network identifier could have the same value as the Kubernetes cluster identifier described in the following section and both aspects would benefit from community feedback and input._

Author: Of course, we can.... That’s a great point, and this is exactly the kind of valuable feedback we were hoping to get from Strimzi maintainers. Really appreciate your prompt response as always!

> _Note: This network identifier could have the same value as the Kubernetes cluster identifier described in the following section and both aspects would benefit from community feedback and input._

Author: Revised proposal have addressed this!

A design for better management of Kubernetes clients for remote clusters is required and is dependent upon community input.

**All aspects of user configuration described above would benefit from further discussion within the community.**

## Additional considerations and reference information

#### Example of multi-cluster advertised.listener and controller.quorum.voters

The Operator configures `advertised.listeners` and `controller.quorum.voters` to support cross-cluster communication in a stretch Kafka deployment.

**Default `advertised.listeners` format**

In a regular (single Kubernetes cluster) deployment, the `advertised.listeners` configuration uses this format:

```
advertised.listeners=REPLICATION-9091://<broker-pod-name>.<broker-service-name>.<namespace>.svc:9091,
PLAIN-9092://<broker-pod-name>.<broker-service-name>.<namespace>.svc:9092,
TLS-9093://<broker-pod-name>.<broker-service-name>.<namespace>.svc:9093
```

**Modified format for a stretch Kafka cluster**

> When Kafka is deployed using "stretch mode", the operator modifies `advertised.listeners` to include a network identifier, such as the Submariner or Cilium `cluster-id`. The modified format is:

Reviewer 8: Yes, the modified format works because the actual routing is being handled by the cloud native network technology like Cilium. 

This will be how the operator will setup the Kafka config. When creating a stretch cluster the cluster operator will set advertised listeners and controller quorum voters in this format instead of the default used today.

The central Cluster Operator replicates Kafka, KafkaNodePool across remote clusters as backup copies.
These CRs are annotated as inactive (e.g., strimzi.io/standby=true) so that remote clusters do not reconcile them.
In case of failure, users remove the standby annotation and update the CO deployment to promote the cluster to central.
This accelerates recovery but adds complexity in keeping replicated CRs up to date.

##### 4. Automated Failover (Future Consideration)
The system detects central cluster failure and automatically elects a new primary cluster.
Replicated CRs become active, and the CO resumes reconciliation from the new cluster.
This offers zero manual intervention but requires leader election, coordination, and consistent state replication, which are non-trivial.

#### Recommended Approach
As a starting point, we recommend manual recovery (Option 1) and GitOps-based recovery (Option 2) due to their simplicity and lower implementation overhead.
We do not plan to implement automated failover (Option 4) at this stage. However, CR replication (Option 3) remains an option if users demand a faster failover mechanism.

#### Entity operator

> We would recommend that all KafkaTopic and KafkaUser resources are managed from the cluster that holds Kafka and KafkaNodePool resources, and that should be the cluster where the entity operator should be enabled.

Reviewer 5: Have you put any thought into what this means for clients running in one of the non-central clusters? It doesn't support it today, but one option would be to use Kafka Access Operator as we have an open issue to support watching Kafka clusters in other Kubernetes clusters.

Author: In a stretched cluster setup with the Entity Operator in the central cluster, existing connected clients will continue to work normally even if the central cluster goes down. This includes producers, consumers, and other clients already authenticated and authorized.

Clients using TLS or SCRAM will keep working as long as their credentials (certificates/passwords) remain valid. TLS clients may face issues only if certificates expire during the outage. SCRAM credentials, stored in Kafka, are unaffected.

Existing topics and their data remain fully accessible. Brokers in surviving clusters retain all metadata (thanks to KRaft), so no disruption occurs for live workloads.

New clients in remote clusters will fail if they rely on KafkaUser Secrets that weren’t pre-synced. Similarly, creating/modifying topics or users via CRs won’t work until the central cluster recovers.

TBH we haven't look too much into the Access Operator. AFAIU it could simplify multi cluster auth, but for now, manual Secret sync or external tools (like Vault....) are sufficient workarounds.

```yaml
- name: STRIMZI_POD_SET_RECONCILIATION_ONLY
  value: true
```

This configuration provides important safeguards:

- It minimizes the resources (CPU and memory) required by the operator on remote Kubernetes clusters.
- It simplifies the operator logs by reducing noise.
- It prevents accidental reconciliation of other Strimzi resources (such as Kafka, KafkaNodePool, or KafkaConnect) if they are mistakenly created in the remote cluster.

It is therefore essential to set `STRIMZI_POD_SET_RECONCILIATION_ONLY=true` when deploying the operator to remote clusters in a stretch cluster deployment.
The official Strimzi documentation should clearly state that it is necessary to set STRIMZI_POD_SET_RECONCILIATION_ONLY=true when deploying the operator to remote clusters.

#### Step 3: Create Kafka and KafkaNodePool resources in the central cluster

##### Kafka CR

To enable stretch cluster functionality, users must explicitly opt in by adding the following annotation to the Kafka resource:

```yaml
metadata:
  annotations:

>     strimzi.io/enable-stretch-cluster: "true"

Author: Yes, we can change that if you think it's less verbose and still conveys the purpose

##### KafkaNodePool CR

The `KafkaNodePool` CR remains in the central cluster but must include a new annotation to indicate the target Kubernetes cluster where the corresponding Kafka node resources should be deployed.
Based on this annotation, the Cluster Operator will create the associated resources — including StrimziPodSet, Services, Secrets, ConfigMaps, and others — in the specified remote cluster.
The `KafkaNodePool` itself is never deployed outside the central cluster.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaNodePool
metadata:
  annotations:
    strimzi.io/stretch-cluster-alias: "cluster3"
```

The identifier used here is required and must match one of the values defined by the user in 'Step 1' and added to the environment variable map value in 'Step 2'.
This annotation is used to construct critical Kafka configuration properties such as `controller.quorum.voters` and `advertised.listeners`.

> If the identifier does not match any of the values defined in the `STRIMZI_REMOTE_KUBE_CONFIG` environment variable, the operator will deploy the node pool to the central Kubernetes cluster.

Author: Yeah , I agree I have pushed this commit https://github.com/strimzi/proposals/pull/129/commits/00055c69f7a322f45022c850bca06e14ef5198b6

to address this

| **Kafka Broker Certificates Secret**         | ✅                   |                            |                            |
| **Kafka Broker Service**                     | ✅                   |                            |                            |
| **Broker Pods**                              |                      |                            | ✅                          |
| **Controller Pods**                           |                      |                            | ✅                          |
| **Persistent Volume Claims (PVCs) for storage** |                      | ✅                          |                            |
| **Pod Disruption Budget**                    | ✅                   |                             |                            |

In a Stretch Kafka deployment, the operator on the central cluster must ensure that all resources derived from the Kafka CR and KafkaNodePool CR are also replicated in the remote clusters, including Persistent Volume Claims (PVCs) for storage.

In addition, the cluster operator normally creates network policies to restrict traffic from outside the namespace.

> To enable pod-to-pod communication across Kubernetes clusters, the cluster operator will not create any network policies when configured to deploy a stretch cluster.

Author: You're absolutely right, and thank you for catching that — it's a great point. The statement about not creating network policies in stretch mode was outdated and contradicts what we've correctly addressed in another part of the proposal. In reality, the default NetworkPolicy that Strimzi creates should not pose a problem, as the underlying MCS solution (e.g., Submariner, Cilium ClusterMesh) is responsible for ensuring inter-cluster pod connectivity and honoring network policy semantics across clusters.

This was a miss on our part — thanks again for pointing it out! I've now updated the proposal to remove the incorrect statement from this section for consistency.

Finalizers will be added by default to the `Kafka` and `KafkaNodePool` resources in the central cluster to ensure the operator removes related remote cluster resources.
A user can disable the finalizers by setting the `STRIMZI_USE_FINALIZER` environment variable to `false`.

#### Rack awareness in stretch clusters
For Kubernetes clusters that are not zone aware, the worker nodes across all Kubernetes clusters involved must be labelled with a zone alias.
The zone alias cannot be the same as the `stretch-cluster-alias`.
Replicas are distributed across multiple Kubernetes clusters based on topology labels across the stretched Kafka cluster.
Broker `ConfigMaps` will be updated to contain rack configuration as they would for a single Kubernetes cluster deployment.

#### Disaster recovery; Handling central cluster failure

In a stretch Kafka deployment, the central Kubernetes cluster manages Kafka resources across multiple clusters, but Kafka brokers and controllers continue operating even if the central cluster fails.
The key challenge is restoring the administrative control plane while ensuring minimal downtime for Kafka clients.

###### Recovery approaches

> There are several possible recovery methods:

Reviewer 3: ```suggestion

There are several possible recovery methods:
```

**1. Manual recovery (baseline approach)**

> Users bring up a new Kubernetes cluster and manually apply the `Kafka` and `KafkaNodePool` CRs.

Reviewer 3: ```suggestion

Users bring up a new Kubernetes cluster and manually apply the `Kafka` and `KafkaNodePool` CRs.
```

Alternatively, an existing remote cluster can be promoted to central by applying these CRs and updating the cluster operator deployment.
This approach is simple and reliable but requires user intervention.

**2. GitOps based recovery**

Users store CRs in GitOps tools (e.g., ArgoCD, Flux).
When a new central cluster is created, GitOps tools automatically reconcile the resources.
This reduces manual effort but requires GitOps workflows to be set up in advance.

#### Feature Enablement and Rollout Plan

> The Stretch Kafka Cluster support will be introduced as an optional feature behind a feature gate called `UseStretchCluster`.

Reviewer 3: ```suggestion

The Stretch Kafka Cluster support will be introduced as an optional feature behind a feature gate called `UseStretchCluster`.
```

The maturity and rollout of this feature gate will follow the standard Strimzi process:
- The feature gate will be disabled by default, allowing early adopters and community members to safely test the functionality without affecting production environments.
- After at least two Strimzi releases, and based on user feedback and observed stability, enabling the feature gate by default may be considered.

Existing Kafka cluster deployments will remain unaffected unless users explicitly enable the UseStretchCluster feature gate and configure the necessary settings.

## Affected/not affected projects

This proposal only impacts the strimzi-kafka-operator project.

### Entity operator

> To maintain the central cluster as the single control plane, all `KafkaTopic` and `KafkaUser` resources must be managed from the central cluster; This is also where the entity operator will be deployed.

Reviewer 3: ```suggestion

To maintain the central cluster as the single control plane, all `KafkaTopic` and `KafkaUser` resources must be managed from the central cluster; This is also where the entity operator will be deployed.
```

The expectation is that the entity operator will not be impacted any further by changes made to support a stretch cluster.

### Kafka Connect, Kafka Bridge and MirrorMaker2

> This proposal does not cover stretching Kafka Connect, Kafka MirrorMaker 2 or the Kafka Bridge.

Reviewer 3: ```suggestion

This proposal does not cover stretching Kafka Connect, Kafka MirrorMaker 2 or the Kafka Bridge.
```

These components will be deployed to the central cluster and will function as they do today.
Operators running in remote clusters will not manage KafkaBridge, KafkaConnect, KafkaConnector, or KafkaMirrorMaker2 resources.

### Drain Cleaner

> This proposal does not affect Drain Cleaner.

Reviewer 3: ```suggestion

This proposal does not affect Drain Cleaner.
```

##### 3. CR Replication with Manual Failover

The central Cluster Operator replicates Kafka, KafkaNodePool across remote clusters as backup copies.
These CRs are annotated as inactive (e.g., strimzi.io/standby=true) so that remote clusters do not reconcile them.
In case of failure, users remove the standby annotation and update the CO deployment to promote the cluster to central.
This accelerates recovery but adds complexity in keeping replicated CRs up to date.

##### 4. Automated Failover (Future Consideration)
The system detects central cluster failure and automatically elects a new primary cluster.
Replicated CRs become active, and the CO resumes reconciliation from the new cluster.
This offers zero manual intervention but requires leader election, coordination, and consistent state replication, which are non-trivial.

#### Recommended Approach
As a starting point, we recommend manual recovery (Option 1) and GitOps-based recovery (Option 2) due to their simplicity and lower implementation overhead.
We do not plan to implement automated failover (Option 4) at this stage. However, CR replication (Option 3) remains an option if users demand a faster failover mechanism.

> 

Reviewer 7: This looks good, but I would use the "rejected alternatives" section to list the approaches that are out of scope for this proposal, but still reported as possible future enhancement (3 and 4).

Author: updated this section ... I have added a new `Rejected Alternatives` section in the proposal
